// @google/jules-sdk/dist/index.mjs
import * as path from "node:path";
import { join } from "node:path";
import { homedir } from "node:os";
import { existsSync, accessSync, constants } from "node:fs";
import * as fs from "fs/promises";
import * as path$1 from "path";
import { createReadStream, createWriteStream } from "fs";
import * as readline from "readline";
import { writeFile, readFile, rm } from "node:fs/promises";
import { Buffer as Buffer$1 } from "node:buffer";
import { setTimeout as setTimeout$1 } from "node:timers/promises";
import * as crypto from "node:crypto";
class JulesError extends Error {
  /** The original error that caused this error, if any. */
  cause;
  constructor(message, options) {
    super(message);
    this.name = this.constructor.name;
    this.cause = options?.cause;
  }
}
class JulesNetworkError extends JulesError {
  url;
  constructor(url, options) {
    super(`Network request to ${url} failed`, options);
    this.url = url;
  }
}
class JulesApiError extends JulesError {
  url;
  status;
  statusText;
  constructor(url, status, statusText, message, options) {
    const finalMessage = message ?? `[${status} ${statusText}] Request to ${url} failed`;
    super(finalMessage, options);
    this.url = url;
    this.status = status;
    this.statusText = statusText;
  }
}
class JulesAuthenticationError extends JulesApiError {
  constructor(url, status, statusText) {
    super(
      url,
      status,
      statusText,
      `[${status} ${statusText}] Authentication to ${url} failed. Ensure your API key is correct.`
    );
  }
}
class JulesRateLimitError extends JulesApiError {
  constructor(url, status, statusText) {
    super(
      url,
      status,
      statusText,
      `[${status} ${statusText}] API rate limit exceeded for ${url}.`
    );
  }
}
class MissingApiKeyError extends JulesError {
  constructor() {
    super(
      "Jules API key is missing. Pass it to the constructor or set the JULES_API_KEY environment variable."
    );
  }
}
class SourceNotFoundError extends JulesError {
  constructor(sourceIdentifier) {
    super(`Could not get source '${sourceIdentifier}'`);
  }
}
class AutomatedSessionFailedError extends JulesError {
  constructor(reason) {
    let message = "The Jules automated session terminated with a FAILED state.";
    if (reason) {
      message += ` Reason: ${reason}`;
    }
    super(message);
  }
}
class TimeoutError extends JulesError {
  constructor(message) {
    super(message);
  }
}
class SyncInProgressError extends JulesError {
  constructor() {
    super(
      "A sync operation is already in progress. Wait for it to complete before starting another."
    );
  }
}
class InvalidStateError extends JulesError {
  constructor(message) {
    super(message);
  }
}
class ApiClient {
  apiKey;
  baseUrl;
  requestTimeoutMs;
  rateLimitConfig;
  semaphore;
  constructor(options) {
    this.apiKey = options.apiKey;
    this.baseUrl = options.baseUrl;
    this.requestTimeoutMs = options.requestTimeoutMs;
    this.rateLimitConfig = {
      maxRetryTimeMs: options.rateLimitRetry?.maxRetryTimeMs ?? 3e5,
      // 5 minutes
      baseDelayMs: options.rateLimitRetry?.baseDelayMs ?? 1e3,
      maxDelayMs: options.rateLimitRetry?.maxDelayMs ?? 3e4
    };
    this.semaphore = new Semaphore(options.maxConcurrentRequests ?? 50);
  }
  async request(endpoint, options = {}) {
    const {
      method = "GET",
      body,
      query,
      headers: customHeaders,
      _isRetry
    } = options;
    const url = this.resolveUrl(endpoint);
    if (query) {
      Object.entries(query).forEach(([key, value]) => {
        url.searchParams.append(key, String(value));
      });
    }
    const headers = {
      "Content-Type": "application/json",
      ...customHeaders
    };
    if (this.apiKey) {
      headers["X-Goog-Api-Key"] = this.apiKey;
    } else {
      throw new MissingApiKeyError();
    }
    let response;
    try {
      await this.semaphore.acquire();
      response = await this.fetchWithTimeout(url.toString(), {
        method,
        headers,
        body: body ? JSON.stringify(body) : void 0
      });
    } finally {
      this.semaphore.release();
    }
    if (!response.ok) {
      if (response.status === 429 || [500, 502, 503, 504].includes(response.status)) {
        const startTime = options._rateLimitStartTime || Date.now();
        const elapsed = Date.now() - startTime;
        const retryCount = options._rateLimitRetryCount || 0;
        if (elapsed < this.rateLimitConfig.maxRetryTimeMs) {
          const rawDelay = this.rateLimitConfig.baseDelayMs * Math.pow(2, retryCount);
          const delay = Math.min(rawDelay, this.rateLimitConfig.maxDelayMs);
          await new Promise((resolve) => setTimeout(resolve, delay));
          return this.request(endpoint, {
            ...options,
            _rateLimitStartTime: startTime,
            _rateLimitRetryCount: retryCount + 1
          });
        }
        if (response.status === 429) {
          throw new JulesRateLimitError(
            url.toString(),
            response.status,
            response.statusText
          );
        }
      }
      switch (response.status) {
        case 401:
        case 403:
          throw new JulesAuthenticationError(
            url.toString(),
            response.status,
            response.statusText
          );
        default:
          const errorBody = await response.text().catch(() => "Could not read error body");
          const message = `[${response.status} ${response.statusText}] ${method} ${url.toString()} - ${errorBody}`;
          throw new JulesApiError(
            url.toString(),
            response.status,
            response.statusText,
            message
          );
      }
    }
    const responseText = await response.text();
    if (!responseText) {
      return {};
    }
    return JSON.parse(responseText);
  }
  resolveUrl(path2) {
    return new URL(`${this.baseUrl}/${path2}`);
  }
  async fetchWithTimeout(url, opts) {
    const controller = new AbortController();
    const timeoutId = setTimeout(
      () => controller.abort(),
      this.requestTimeoutMs
    );
    try {
      const response = await fetch(url, {
        ...opts,
        signal: controller.signal
      });
      return response;
    } catch (error) {
      throw new JulesNetworkError(url, {
        cause: error
      });
    } finally {
      clearTimeout(timeoutId);
    }
  }
}
class Semaphore {
  constructor(maxConcurrentRequests) {
    this.maxConcurrentRequests = maxConcurrentRequests;
  }
  currentRequests = 0;
  queue = [];
  async acquire() {
    if (this.currentRequests < this.maxConcurrentRequests) {
      this.currentRequests++;
      return Promise.resolve();
    }
    return new Promise((resolve) => {
      this.queue.push(resolve);
    });
  }
  release() {
    if (this.queue.length > 0) {
      const resolve = this.queue.shift();
      if (resolve) {
        resolve();
      }
    } else {
      this.currentRequests--;
    }
  }
}
function mapRawSourceToSdkSource(rawSource) {
  if (rawSource.githubRepo) {
    const { defaultBranch, branches, ...rest } = rawSource.githubRepo;
    return {
      name: rawSource.name,
      id: rawSource.id,
      type: "githubRepo",
      githubRepo: {
        ...rest,
        defaultBranch: defaultBranch?.displayName,
        branches: branches?.map((b) => b.displayName)
      }
    };
  }
  throw new Error(`Unknown source type for source: ${rawSource.name}`);
}
class SourceManagerImpl {
  apiClient;
  constructor(apiClient) {
    this.apiClient = apiClient;
  }
  /**
   * Lists all connected sources.
   *
   * **Logic:**
   * - Automatically handles API pagination by following `nextPageToken`.
   * - Yields sources one by one as they are retrieved.
   */
  async *list(options = {}) {
    let pageToken = void 0;
    while (true) {
      const params = {
        pageSize: (options.pageSize || 100).toString()
      };
      if (options.filter) {
        params.filter = options.filter;
      }
      if (pageToken) {
        params.pageToken = pageToken;
      }
      const response = await this.apiClient.request(
        "sources",
        { query: params }
      );
      if (response && response.sources) {
        for (const rawSource of response.sources) {
          yield mapRawSourceToSdkSource(rawSource);
        }
      }
      pageToken = response?.nextPageToken;
      if (!pageToken) {
        break;
      }
    }
  }
  /**
   * Retrieves a specific source by its external identifier.
   *
   * **Data Transformation:**
   * - Constructs a resource name (e.g., `sources/github/owner/repo`) from the input filter.
   *
   * @param filter Filter criteria (currently supports GitHub repo name).
   * @returns The matching Source object, or `undefined` if not found (404).
   * @throws {Error} If the filter format is invalid.
   */
  async get(filter) {
    const { github } = filter;
    if (!github || !github.includes("/")) {
      throw new Error("Invalid GitHub filter. Expected format: 'owner/repo'.");
    }
    const resourceName = `sources/github/${github}`;
    try {
      const rawSource = await this.apiClient.request(resourceName);
      if (!rawSource) {
        return void 0;
      }
      return mapRawSourceToSdkSource(rawSource);
    } catch (error) {
      if (error instanceof JulesApiError && error.status === 404) {
        return void 0;
      }
      throw error;
    }
  }
}
function createSourceManager(apiClient) {
  const manager = new SourceManagerImpl(apiClient);
  const callable = manager.list.bind(manager);
  const sourceManager = callable;
  sourceManager.get = manager.get.bind(manager);
  return sourceManager;
}
function isWritable(dir) {
  try {
    accessSync(dir, constants.W_OK);
    return true;
  } catch {
    return false;
  }
}
function getRootDir() {
  const julesHome = process.env.JULES_HOME;
  if (julesHome && isWritable(julesHome)) {
    return julesHome;
  }
  const cwd = process.cwd();
  const isInProject = existsSync(path.join(cwd, "package.json"));
  if (isInProject && cwd !== "/" && isWritable(cwd)) {
    return cwd;
  }
  const home = process.env.HOME;
  if (home && home !== "/" && isWritable(home)) {
    return home;
  }
  const osHome = homedir();
  if (osHome && osHome !== "/" && isWritable(osHome)) {
    return osHome;
  }
  const tmpDir = process.env.TMPDIR || process.env.TMP || "/tmp";
  return tmpDir;
}
function parseUnidiff(patch) {
  if (!patch) return [];
  const files = [];
  const diffSections = patch.split(/^diff --git /m).filter(Boolean);
  for (const section of diffSections) {
    const lines = section.split("\n");
    let path2 = "";
    let fromPath = "";
    let toPath = "";
    for (const line of lines) {
      if (line.startsWith("--- ")) {
        fromPath = line.slice(4).replace(/^a\//, "").replace(/^\/dev\/null$/, "");
      } else if (line.startsWith("+++ ")) {
        toPath = line.slice(4).replace(/^b\//, "").replace(/^\/dev\/null$/, "");
      }
    }
    let changeType;
    if (fromPath === "" || lines.some((l) => l.startsWith("--- /dev/null"))) {
      changeType = "created";
      path2 = toPath;
    } else if (toPath === "" || lines.some((l) => l.startsWith("+++ /dev/null"))) {
      changeType = "deleted";
      path2 = fromPath;
    } else {
      changeType = "modified";
      path2 = toPath;
    }
    if (!path2) continue;
    let additions = 0;
    let deletions = 0;
    let inHunk = false;
    for (const line of lines) {
      if (line.startsWith("@@")) {
        inHunk = true;
        continue;
      }
      if (inHunk) {
        if (line.startsWith("+") && !line.startsWith("+++")) {
          additions++;
        } else if (line.startsWith("-") && !line.startsWith("---")) {
          deletions++;
        }
      }
    }
    files.push({ path: path2, changeType, additions, deletions });
  }
  return files;
}
function parseUnidiffWithContent(patch) {
  if (!patch) return [];
  const files = [];
  const diffSections = patch.split(/^diff --git /m).filter(Boolean);
  for (const section of diffSections) {
    const lines = section.split("\n");
    let path2 = "";
    let fromPath = "";
    let toPath = "";
    for (const line of lines) {
      if (line.startsWith("--- ")) {
        fromPath = line.slice(4).replace(/^a\//, "").replace(/^\/dev\/null$/, "");
      } else if (line.startsWith("+++ ")) {
        toPath = line.slice(4).replace(/^b\//, "").replace(/^\/dev\/null$/, "");
      }
    }
    let changeType;
    if (fromPath === "" || lines.some((l) => l.startsWith("--- /dev/null"))) {
      changeType = "created";
      path2 = toPath;
    } else if (toPath === "" || lines.some((l) => l.startsWith("+++ /dev/null"))) {
      changeType = "deleted";
      path2 = fromPath;
    } else {
      changeType = "modified";
      path2 = toPath;
    }
    if (!path2) continue;
    let additions = 0;
    let deletions = 0;
    let inHunk = false;
    const contentLines = [];
    for (const line of lines) {
      if (line.startsWith("@@")) {
        inHunk = true;
        continue;
      }
      if (inHunk) {
        if (line.startsWith("+") && !line.startsWith("+++")) {
          additions++;
          contentLines.push(line.slice(1));
        } else if (line.startsWith("-") && !line.startsWith("---")) {
          deletions++;
        }
      }
    }
    const content = changeType === "deleted" ? "" : contentLines.join("\n");
    files.push({ path: path2, changeType, content, additions, deletions });
  }
  return files;
}
function createGeneratedFiles(files) {
  return {
    all: () => files,
    get: (path2) => files.find((f) => f.path === path2),
    filter: (changeType) => files.filter((f) => f.changeType === changeType)
  };
}
class MediaArtifact {
  type = "media";
  data;
  format;
  platform;
  activityId;
  constructor(artifact, platform, activityId) {
    this.data = artifact.data;
    this.format = artifact.mimeType;
    this.platform = platform;
    this.activityId = activityId;
  }
  /**
   * Saves the media artifact to a file.
   *
   * **Side Effects:**
   * - Node.js: Writes the file to disk (overwrites if exists).
   * - Browser: Saves the file to the 'artifacts' object store in IndexedDB.
   *
   * @param filepath The path where the file should be saved.
   */
  async save(filepath) {
    await this.platform.saveFile(
      filepath,
      this.data,
      "base64",
      this.activityId
    );
  }
  /**
   * Converts the media artifact to a data URL.
   * Useful for displaying images in a browser.
   *
   * **Data Transformation:**
   * - Prefixes the base64 data with `data:<mimeType>;base64,`.
   *
   * @returns A valid Data URI string.
   */
  toUrl() {
    return this.platform.createDataUrl(this.data, this.format);
  }
}
class BashArtifact {
  type = "bashOutput";
  command;
  stdout;
  stderr;
  exitCode;
  constructor(artifact) {
    this.command = artifact.command;
    this.stdout = artifact.output;
    this.stderr = "";
    this.exitCode = artifact.exitCode;
  }
  /**
   * Formats the bash output as a string, mimicking a terminal session.
   *
   * **Data Transformation:**
   * - Combines `stdout` and `stderr`.
   * - Formats the command with a `$` prompt.
   * - Appends the exit code.
   */
  toString() {
    const output = [this.stdout, this.stderr].filter(Boolean).join("");
    const commandLine = `$ ${this.command}`;
    const outputLine = output ? `${output}
` : "";
    const exitLine = `[exit code: ${this.exitCode ?? "N/A"}]`;
    return `${commandLine}
${outputLine}${exitLine}`;
  }
}
class ChangeSetArtifact {
  type = "changeSet";
  source;
  gitPatch;
  constructor(source, gitPatch) {
    this.source = source;
    this.gitPatch = gitPatch;
  }
  /**
   * Parses the unified diff and returns structured file change information.
   *
   * **Data Transformation:**
   * - Extracts file paths from diff headers.
   * - Determines change type (created/modified/deleted) from /dev/null markers.
   * - Counts additions (+) and deletions (-) in hunks.
   *
   * @returns Parsed diff with file paths, change types, and line counts.
   */
  parsed() {
    if (!this.gitPatch?.unidiffPatch) {
      return {
        files: [],
        summary: { totalFiles: 0, created: 0, modified: 0, deleted: 0 }
      };
    }
    const files = parseUnidiff(this.gitPatch.unidiffPatch);
    const summary = {
      totalFiles: files.length,
      created: files.filter((f) => f.changeType === "created").length,
      modified: files.filter((f) => f.changeType === "modified").length,
      deleted: files.filter((f) => f.changeType === "deleted").length
    };
    return { files, summary };
  }
}
function mapRestArtifactToSdkArtifact(restArtifact, platform, activityId) {
  if ("changeSet" in restArtifact) {
    return new ChangeSetArtifact(
      restArtifact.changeSet.source,
      restArtifact.changeSet.gitPatch
    );
  }
  if ("media" in restArtifact) {
    return new MediaArtifact(restArtifact.media, platform, activityId);
  }
  if ("bashOutput" in restArtifact) {
    return new BashArtifact(restArtifact.bashOutput);
  }
  throw new Error(`Unknown artifact type: ${JSON.stringify(restArtifact)}`);
}
function mapRestActivityToSdkActivity(restActivity, platform) {
  const {
    name,
    createTime,
    originator,
    artifacts: rawArtifacts,
    description
  } = restActivity;
  const activityId = name.split("/").pop();
  const artifacts = (rawArtifacts || []).map(
    (artifact) => mapRestArtifactToSdkArtifact(artifact, platform, activityId)
  );
  const baseActivity = {
    name,
    id: activityId,
    description,
    createTime,
    originator: originator || "system",
    artifacts
  };
  if (restActivity.agentMessaged) {
    return {
      ...baseActivity,
      type: "agentMessaged",
      message: restActivity.agentMessaged.agentMessage
    };
  }
  if (restActivity.userMessaged) {
    return {
      ...baseActivity,
      type: "userMessaged",
      message: restActivity.userMessaged.userMessage
    };
  }
  if (restActivity.planGenerated) {
    return {
      ...baseActivity,
      type: "planGenerated",
      plan: restActivity.planGenerated.plan
    };
  }
  if (restActivity.planApproved) {
    return {
      ...baseActivity,
      type: "planApproved",
      planId: restActivity.planApproved.planId
    };
  }
  if (restActivity.progressUpdated) {
    return {
      ...baseActivity,
      type: "progressUpdated",
      title: restActivity.progressUpdated.title,
      description: restActivity.progressUpdated.description
    };
  }
  if (restActivity.sessionCompleted) {
    return {
      ...baseActivity,
      type: "sessionCompleted"
    };
  }
  if (restActivity.sessionFailed) {
    return {
      ...baseActivity,
      type: "sessionFailed",
      reason: restActivity.sessionFailed.reason
    };
  }
  throw new Error("Unknown activity type");
}
function mapRestStateToSdkState(state) {
  switch (state) {
    case "STATE_UNSPECIFIED":
      return "unspecified";
    case "QUEUED":
      return "queued";
    case "PLANNING":
      return "planning";
    case "AWAITING_PLAN_APPROVAL":
      return "awaitingPlanApproval";
    case "AWAITING_USER_FEEDBACK":
      return "awaitingUserFeedback";
    case "IN_PROGRESS":
      return "inProgress";
    case "PAUSED":
      return "paused";
    case "FAILED":
      return "failed";
    case "COMPLETED":
      return "completed";
    default:
      return "unspecified";
  }
}
function mapRestSourceToSdkSource(rest) {
  if (rest.githubRepo) {
    const { defaultBranch, branches, ...other } = rest.githubRepo;
    return {
      type: "githubRepo",
      name: rest.name,
      id: rest.id,
      githubRepo: {
        ...other,
        defaultBranch: defaultBranch?.displayName,
        branches: branches?.map((b) => b.displayName)
      }
    };
  }
  throw new Error(`Unknown source type: ${JSON.stringify(rest)}`);
}
function mapRestOutputToSdkOutput(rest) {
  if (rest.pullRequest) {
    return {
      type: "pullRequest",
      pullRequest: rest.pullRequest
    };
  }
  if (rest.changeSet) {
    return {
      type: "changeSet",
      changeSet: rest.changeSet
    };
  }
  throw new Error(`Unknown output type: ${JSON.stringify(rest)}`);
}
function mapRestSessionToSdkSession(rest, platform) {
  const session = {
    ...rest,
    archived: rest.archived ?? false,
    state: mapRestStateToSdkState(rest.state),
    requirePlanApproval: rest.requirePlanApproval,
    automationMode: rest.automationMode,
    outputs: (rest.outputs || []).map(mapRestOutputToSdkOutput),
    source: rest.source ? mapRestSourceToSdkSource(rest.source) : void 0,
    generatedFiles: rest.generatedFiles,
    activities: void 0,
    outcome: void 0
  };
  if (rest.activities && platform) {
    session.activities = rest.activities.map(
      (a) => mapRestActivityToSdkActivity(a, platform)
    );
  }
  try {
    session.outcome = mapSessionResourceToOutcome(session);
  } catch (error) {
    if (error instanceof AutomatedSessionFailedError) {
      session.outcome = {
        sessionId: session.id,
        title: session.title,
        state: "failed",
        outputs: session.outputs,
        generatedFiles: () => createGeneratedFiles([]),
        changeSet: () => void 0
      };
    } else {
      throw error;
    }
  }
  return session;
}
function mapSessionResourceToOutcome(session) {
  if (session.state === "failed") {
    throw new AutomatedSessionFailedError(`Session ${session.id} failed.`);
  }
  const outputs = session.outputs ?? [];
  const prOutput = outputs.find((o) => "pullRequest" in o);
  const pullRequest = prOutput ? prOutput.pullRequest : void 0;
  const changeSetOutput = outputs.find((o) => "changeSet" in o);
  const changeSet = changeSetOutput ? changeSetOutput.changeSet : void 0;
  return {
    sessionId: session.id,
    title: session.title,
    state: "completed",
    // We only call this mapper on a completed session.
    pullRequest,
    outputs,
    generatedFiles: () => {
      if (!changeSet?.gitPatch?.unidiffPatch) {
        return createGeneratedFiles([]);
      }
      const files = parseUnidiffWithContent(changeSet.gitPatch.unidiffPatch);
      return createGeneratedFiles(files);
    },
    changeSet: () => {
      if (!changeSet?.gitPatch) {
        return void 0;
      }
      return new ChangeSetArtifact("session", changeSet.gitPatch);
    }
  };
}
const sleep$1 = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
const DEFAULT_INITIAL_RETRIES = 10;
const MAX_RETRY_DELAY_MS = 3e4;
async function* streamActivities(sessionId, apiClient, pollingInterval, platform, options = {}) {
  let pageToken = void 0;
  let isFirstCall = true;
  let lastSeenTime = "";
  const seenIdsAtLastTime = /* @__PURE__ */ new Set();
  while (true) {
    let response;
    try {
      response = await apiClient.request(
        `sessions/${sessionId}/activities`,
        {
          query: {
            pageSize: "50",
            // A reasonable page size
            ...pageToken ? { pageToken } : {}
          }
        }
      );
    } catch (error) {
      if (isFirstCall && error instanceof JulesApiError && error.status === 404) {
        let lastError = error;
        let successfulResponse;
        let delay = 1e3;
        const maxRetries = options.initialRetries ?? DEFAULT_INITIAL_RETRIES;
        for (let i = 0; i < maxRetries; i++) {
          await sleep$1(delay);
          delay = Math.min(delay * 2, MAX_RETRY_DELAY_MS);
          try {
            successfulResponse = await apiClient.request(
              `sessions/${sessionId}/activities`,
              {
                query: {
                  pageSize: "50",
                  ...pageToken ? { pageToken } : {}
                }
              }
            );
            break;
          } catch (retryError) {
            if (retryError instanceof JulesApiError && retryError.status === 404) {
              lastError = retryError;
            } else {
              throw retryError;
            }
          }
        }
        if (successfulResponse) {
          response = successfulResponse;
        } else {
          throw lastError;
        }
      } else {
        throw error;
      }
    }
    isFirstCall = false;
    const activities = response.activities || [];
    for (const rawActivity of activities) {
      const activity = mapRestActivityToSdkActivity(rawActivity, platform);
      if (activity.createTime < lastSeenTime) {
        continue;
      }
      if (activity.createTime === lastSeenTime) {
        if (seenIdsAtLastTime.has(activity.id)) {
          continue;
        }
        seenIdsAtLastTime.add(activity.id);
      } else {
        lastSeenTime = activity.createTime;
        seenIdsAtLastTime.clear();
        seenIdsAtLastTime.add(activity.id);
      }
      if (options.exclude?.originator && activity.originator === options.exclude.originator) {
        continue;
      }
      yield activity;
    }
    if (response.nextPageToken) {
      pageToken = response.nextPageToken;
      continue;
    } else {
      pageToken = void 0;
      await sleep$1(pollingInterval);
    }
  }
}
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
async function pollSession(sessionId, apiClient, predicateFn, pollingInterval, platform, timeoutMs) {
  const startTime = Date.now();
  while (true) {
    const restSession = await apiClient.request(
      `sessions/${sessionId}`
    );
    const session = mapRestSessionToSdkSession(restSession, platform);
    if (predicateFn(session)) {
      return session;
    }
    if (timeoutMs && Date.now() - startTime >= timeoutMs) {
      throw new TimeoutError(
        `Polling for session ${sessionId} timed out after ${timeoutMs}ms`
      );
    }
    await sleep(pollingInterval);
  }
}
async function pollUntilCompletion(sessionId, apiClient, pollingInterval, platform, timeoutMs) {
  return pollSession(
    sessionId,
    apiClient,
    (session) => {
      const state = session.state;
      return state === "completed" || state === "failed";
    },
    pollingInterval,
    platform,
    timeoutMs
  );
}
function isSessionFrozen(lastActivityCreateTime, thresholdDays = 30) {
  const lastActivity = new Date(lastActivityCreateTime);
  const now = /* @__PURE__ */ new Date();
  const ageMs = now.getTime() - lastActivity.getTime();
  const ageDays = ageMs / (1e3 * 60 * 60 * 24);
  return ageDays > thresholdDays;
}
function createTimeFilter(createTime) {
  return `create_time>"${createTime}"`;
}
class DefaultActivityClient {
  constructor(storage, network, platform) {
    this.storage = storage;
    this.network = network;
    this.platform = platform;
  }
  /**
   * Re-hydrates plain artifact objects from storage into rich class instances.
   * JSON serialization loses class information (methods), so we need to restore it.
   *
   * **Behavior:**
   * - Iterates through artifacts in an activity.
   * - If an artifact is a plain object (not a class instance), it's re-instantiated.
   * - Handles backward compatibility: if an artifact is already a class instance, it's skipped.
   *
   * @param activity The activity from storage, potentially with plain artifacts.
   * @returns The same activity with its artifacts guaranteed to be class instances.
   */
  _hydrateActivityArtifacts(activity) {
    if (!activity.artifacts || activity.artifacts.length === 0) {
      return activity;
    }
    const hydratedArtifacts = activity.artifacts.map((artifact) => {
      if (artifact instanceof MediaArtifact) return artifact;
      if (artifact instanceof BashArtifact) return artifact;
      if (artifact instanceof ChangeSetArtifact) return artifact;
      switch (artifact.type) {
        case "changeSet":
          const rawChangeSet = artifact.changeSet || artifact;
          return new ChangeSetArtifact(
            rawChangeSet.source,
            rawChangeSet.gitPatch
          );
        case "bashOutput":
          const rawBashOutput = artifact.bashOutput || artifact;
          return new BashArtifact(rawBashOutput);
        case "media":
          const rawMedia = artifact.media || artifact;
          return new MediaArtifact(rawMedia, this.platform, activity.id);
        default:
          return artifact;
      }
    });
    return {
      ...activity,
      artifacts: hydratedArtifacts
    };
  }
  /**
   * Returns an async iterable of all activities.
   *
   * **Behavior:**
   * - Always syncs new activities from the network first (via hydrate).
   * - Then yields all activities from local storage.
   *
   * This ensures callers always get the complete, up-to-date history
   * rather than potentially stale cached data.
   */
  async *history() {
    await this.hydrate();
    for await (const activity of this.storage.scan()) {
      yield this._hydrateActivityArtifacts(activity);
    }
  }
  /**
   * Syncs new activities from the network to local cache.
   *
   * **Optimization Strategy:**
   * Activities are immutable - once downloaded, they never change.
   * We use the Jules API's pageToken (nanosecond timestamp) to fetch
   * only activities newer than our latest cached one.
   *
   * **Behavior:**
   * - Empty cache: Fetches all activities (no pageToken)
   * - Has cached activities: Constructs pageToken from latest createTime,
   *   fetches only newer activities
   * - Frozen session (> 30 days): Skips API call entirely
   *
   * @returns The number of new activities synced.
   */
  async hydrate() {
    await this.storage.init();
    const latest = await this.storage.latest();
    if (latest?.createTime && isSessionFrozen(latest.createTime)) {
      return 0;
    }
    const filter = latest?.createTime ? createTimeFilter(latest.createTime) : void 0;
    let count = 0;
    let nextPageToken;
    do {
      const response = await this.network.listActivities({
        filter,
        pageToken: nextPageToken
      });
      const existingChecks = await Promise.all(
        response.activities.map((activity) => this.storage.get(activity.id))
      );
      for (let i = 0; i < response.activities.length; i++) {
        const activity = response.activities[i];
        const existing = existingChecks[i];
        if (existing) {
          continue;
        }
        await this.storage.append(activity);
        count++;
      }
      nextPageToken = response.nextPageToken;
    } while (nextPageToken);
    return count;
  }
  /**
   * Returns an async iterable of new activities from the network.
   * This method polls the network and updates the local storage.
   *
   * **Side Effects:**
   * - Polls the network continuously.
   * - Appends new activities to local storage (write-through caching).
   *
   * **Logic:**
   * - Reads the latest activity from storage to determine the "high-water mark".
   * - Ignores incoming activities older than or equal to the high-water mark.
   */
  async *updates() {
    await this.storage.init();
    const latest = await this.storage.latest();
    let highWaterMark = latest?.createTime ? new Date(latest.createTime).getTime() : 0;
    let lastSeenId = latest?.id;
    for await (const activity of this.network.rawStream()) {
      const actTime = new Date(activity.createTime).getTime();
      if (actTime < highWaterMark) {
        continue;
      }
      if (actTime === highWaterMark && activity.id === lastSeenId) {
        continue;
      }
      await this.storage.append(activity);
      highWaterMark = actTime;
      lastSeenId = activity.id;
      yield activity;
    }
  }
  /**
   * Returns a combined stream of history and updates.
   * This is the primary method for consuming the activity stream.
   *
   * **Behavior:**
   * 1. Yields all historical activities from local storage (offline capable).
   * 2. Switches to `updates()` to yield new activities from the network (real-time).
   */
  async *stream() {
    yield* this.history();
    yield* this.updates();
  }
  /**
   * Queries local storage for activities matching the given options.
   */
  async select(options = {}) {
    await this.storage.init();
    const results = [];
    let started = !options.after;
    let count = 0;
    for await (const act of this.storage.scan()) {
      if (!started) {
        if (act.id === options.after) {
          started = true;
        }
        continue;
      }
      if (options.before && act.id === options.before) {
        break;
      }
      if (options.type && act.type !== options.type) {
        continue;
      }
      results.push(this._hydrateActivityArtifacts(act));
      count++;
      if (options.limit && count >= options.limit) {
        break;
      }
    }
    return results;
  }
  /**
   * Lists activities from the network directly.
   * @param options Pagination options.
   */
  async list(options) {
    return this.network.listActivities(options);
  }
  /**
   * Gets a single activity by ID.
   * Implements a "read-through" caching strategy.
   *
   * **Logic:**
   * 1. Checks local storage. If found, returns it immediately (fast).
   * 2. If missing, fetches from the network.
   * 3. Persists the fetched activity to storage (future reads will hit cache).
   * 4. Returns the activity.
   *
   * **Side Effects:**
   * - May perform a network request.
   * - May write to local storage.
   */
  async get(activityId) {
    await this.storage.init();
    const cached = await this.storage.get(activityId);
    if (cached) {
      return this._hydrateActivityArtifacts(cached);
    }
    const fresh = await this.network.fetchActivity(activityId);
    await this.storage.append(fresh);
    return fresh;
  }
}
class NetworkAdapter {
  constructor(apiClient, sessionId, pollingIntervalMs = 5e3, platform) {
    this.apiClient = apiClient;
    this.sessionId = sessionId;
    this.pollingIntervalMs = pollingIntervalMs;
    this.platform = platform;
  }
  /**
   * Fetches a single activity from the API.
   */
  async fetchActivity(activityId) {
    const restActivity = await this.apiClient.request(
      `sessions/${this.sessionId}/activities/${activityId}`
    );
    return mapRestActivityToSdkActivity(restActivity, this.platform);
  }
  /**
   * Lists activities from the API with pagination.
   */
  async listActivities(options) {
    const params = {};
    if (options?.pageSize) {
      params.pageSize = options.pageSize.toString();
    }
    if (options?.pageToken) {
      params.pageToken = options.pageToken;
    }
    if (options?.filter) {
      params.filter = options.filter;
    }
    const response = await this.apiClient.request(`sessions/${this.sessionId}/activities`, { query: params });
    return {
      activities: (response.activities || []).map(
        (activity) => mapRestActivityToSdkActivity(activity, this.platform)
      ),
      nextPageToken: response.nextPageToken
    };
  }
  /**
   * Polls the API for new activities and yields them.
   * This stream never ends unless the process is terminated.
   */
  async *rawStream() {
    while (true) {
      let pageToken = void 0;
      do {
        const response = await this.listActivities({ pageToken });
        for (const activity of response.activities) {
          yield activity;
        }
        pageToken = response.nextPageToken;
      } while (pageToken);
      await this.platform.sleep(this.pollingIntervalMs);
    }
  }
}
const ONE_MONTH_MS = 30 * 24 * 60 * 60 * 1e3;
const ONE_DAY_MS = 24 * 60 * 60 * 1e3;
function determineCacheTier(cached, now = Date.now()) {
  const createdAt = new Date(cached.resource.createTime).getTime();
  const age = now - createdAt;
  const isTerminal = ["failed", "completed"].includes(cached.resource.state);
  if (age > ONE_MONTH_MS) {
    return "frozen";
  }
  const timeSinceSync = now - cached._lastSyncedAt;
  if (isTerminal && timeSinceSync < ONE_DAY_MS) {
    return "warm";
  }
  return "hot";
}
function isCacheValid(cached, now = Date.now()) {
  if (!cached) return false;
  const tier = determineCacheTier(cached, now);
  return tier === "frozen" || tier === "warm";
}
class SessionSnapshotImpl {
  id;
  state;
  url;
  createdAt;
  updatedAt;
  durationMs;
  prompt;
  title;
  pr;
  activities;
  activityCounts;
  timeline;
  insights;
  generatedFiles;
  changeSet;
  constructor(options) {
    const { session, activities = [] } = options.data;
    this.id = session.id;
    this.state = session.state;
    this.url = session.url;
    this.createdAt = new Date(session.createTime);
    this.updatedAt = new Date(session.updateTime);
    this.durationMs = this.updatedAt.getTime() - this.createdAt.getTime();
    this.prompt = session.prompt;
    this.title = session.title;
    if (session.outcome) {
      this.pr = session.outcome.pullRequest;
      this.generatedFiles = session.outcome.generatedFiles();
      this.changeSet = typeof session.outcome.changeSet === "function" ? session.outcome.changeSet : () => void 0;
    } else {
      const prOutput = session.outputs?.find((o) => o.type === "pullRequest");
      this.pr = prOutput?.pullRequest;
      this.generatedFiles = {
        all: () => [],
        get: () => void 0,
        filter: () => []
      };
      this.changeSet = () => void 0;
    }
    this.activities = Object.freeze(activities);
    this.activityCounts = this.computeActivityCounts();
    this.timeline = this.computeTimeline();
    this.insights = this.computeInsights();
    Object.freeze(this);
  }
  computeActivityCounts() {
    const counts = {};
    for (const activity of this.activities) {
      counts[activity.type] = (counts[activity.type] || 0) + 1;
    }
    return counts;
  }
  computeTimeline() {
    return this.activities.map((activity) => ({
      time: activity.createTime,
      type: activity.type,
      summary: this.generateSummary(activity)
    }));
  }
  generateSummary(activity) {
    switch (activity.type) {
      case "planGenerated":
        return `Plan with ${activity.plan.steps.length} steps`;
      case "planApproved":
        return "Plan approved";
      case "sessionCompleted":
        return "Session completed";
      case "sessionFailed":
        return `Failed: ${activity.reason}`;
      case "userMessaged": {
        const msg = activity.message;
        return `User: ${msg.substring(0, 100)}${msg.length > 100 ? "..." : ""}`;
      }
      case "agentMessaged": {
        const msg = activity.message;
        return `Agent: ${msg.substring(0, 100)}${msg.length > 100 ? "..." : ""}`;
      }
      case "progressUpdated": {
        const progress = activity;
        return progress.title || progress.description || "Progress update";
      }
      default:
        return activity.type;
    }
  }
  computeInsights() {
    const failedCommands = this.activities.filter(
      (activity) => activity.artifacts.some((artifact) => {
        if (artifact.type === "bashOutput") {
          return artifact.exitCode !== 0;
        }
        return false;
      })
    );
    return {
      completionAttempts: this.activityCounts["sessionCompleted"] || 0,
      planRegenerations: this.activityCounts["planGenerated"] || 0,
      userInterventions: this.activityCounts["userMessaged"] || 0,
      failedCommands
    };
  }
  toJSON(options = { exclude: ["activities", "generatedFiles"] }) {
    const full = {
      id: this.id,
      state: this.state,
      url: this.url,
      createdAt: this.createdAt.toISOString(),
      updatedAt: this.updatedAt.toISOString(),
      durationMs: this.durationMs,
      prompt: this.prompt,
      title: this.title,
      activities: this.activities,
      activityCounts: this.activityCounts,
      timeline: this.timeline,
      generatedFiles: this.generatedFiles.all(),
      insights: {
        completionAttempts: this.insights.completionAttempts,
        planRegenerations: this.insights.planRegenerations,
        userInterventions: this.insights.userInterventions,
        failedCommandCount: this.insights.failedCommands.length
      },
      pr: this.pr
    };
    if (options?.include) {
      return Object.fromEntries(
        options.include.filter((key) => key in full).map((key) => [key, full[key]])
      );
    }
    if (options?.exclude) {
      const result = { ...full };
      for (const key of options.exclude) {
        delete result[key];
      }
      return result;
    }
    return full;
  }
  toMarkdown() {
    const lines = [];
    lines.push(`# Session: ${this.title}`);
    lines.push(`**Status**: \`${this.state}\` | **ID**: \`${this.id}\``);
    lines.push("");
    lines.push("## Overview");
    lines.push(`- **Duration**: ${Math.round(this.durationMs / 1e3)}s`);
    lines.push(`- **Total Activities**: ${this.activities.length}`);
    if (this.pr) {
      lines.push(`- **Pull Request**: [${this.pr.title}](${this.pr.url})`);
    }
    if (this.generatedFiles.all().length > 0) {
      lines.push(`- **Generated Files**: ${this.generatedFiles.all().length}`);
      for (const file of this.generatedFiles.all()) {
        lines.push(`  - ${file.path}`);
        lines.push(`  - Type: ${file.changeType}`);
        lines.push(`  - Additions: ${file.additions}`);
        lines.push(`  - Deletions: ${file.deletions}`);
      }
    }
    lines.push("");
    lines.push("## Insights");
    lines.push(
      `- **Completion Attempts**: ${this.insights.completionAttempts}`
    );
    lines.push(`- **Plan Regenerations**: ${this.insights.planRegenerations}`);
    lines.push(`- **User Interventions**: ${this.insights.userInterventions}`);
    lines.push(`- **Failed Commands**: ${this.insights.failedCommands.length}`);
    lines.push("");
    lines.push("## Timeline");
    if (this.timeline.length === 0) {
      lines.push("_No activities recorded._");
    } else {
      for (const entry of this.timeline) {
        lines.push(`- **[${entry.type}]** ${entry.summary} _(${entry.time})_`);
      }
    }
    lines.push("");
    if (Object.keys(this.activityCounts).length > 0) {
      lines.push("## Activity Counts");
      lines.push("```");
      for (const [type, count] of Object.entries(this.activityCounts)) {
        lines.push(`${type.padEnd(20)}: ${count}`);
      }
      lines.push("```");
    }
    return lines.join("\n");
  }
}
async function collectAsync(iterable) {
  const items = [];
  for await (const item of iterable) {
    items.push(item);
  }
  return items;
}
class SessionClientImpl {
  id;
  apiClient;
  config;
  sessionStorage;
  // Added property
  _activities;
  platform;
  /**
   * Creates a new instance of SessionClientImpl.
   *
   * @param sessionId The ID of the session.
   * @param apiClient The API client to use for network requests.
   * @param config The configuration options.
   * @param activityStorage The storage engine for activities.
   * @param sessionStorage The storage engine for sessions.
   * @param platform The platform adapter.
   */
  constructor(sessionId, apiClient, config, activityStorage, sessionStorage, platform) {
    this.id = sessionId.replace(/^sessions\//, "");
    this.apiClient = apiClient;
    this.config = config;
    this.sessionStorage = sessionStorage;
    this.platform = platform;
    const network = new NetworkAdapter(
      this.apiClient,
      this.id,
      this.config.pollingIntervalMs,
      platform
    );
    this._activities = new DefaultActivityClient(
      activityStorage,
      network,
      platform
    );
  }
  // Private helper wrapper to enforce resume context
  async request(path2, options = {}) {
    return this.apiClient.request(path2, options);
  }
  /**
   * COLD STREAM: Yields all known past activities from local storage.
   * If local cache is empty, fetches from network first.
   */
  history() {
    return this._activities.history();
  }
  /**
   * Forces a full sync of activities from the network to local cache.
   * @returns The number of new activities synced.
   */
  hydrate() {
    return this._activities.hydrate();
  }
  /**
   * HOT STREAM: Yields ONLY future activities as they arrive from the network.
   */
  updates() {
    return this._activities.updates();
  }
  /**
   * LOCAL QUERY: Performs rich filtering against local storage only.
   *
   * @deprecated Use `session.activities.select()` instead.
   */
  select(options) {
    return this._activities.select(options);
  }
  /**
   * Scoped access to activity-specific operations.
   */
  get activities() {
    return this._activities;
  }
  /**
   * Provides a real-time stream of activities for the session.
   *
   * @param options Options to control the stream.
   */
  async *stream(options = {}) {
    for await (const activity of this._activities.stream()) {
      if (options.exclude?.originator && activity.originator === options.exclude.originator) {
        continue;
      }
      yield activity;
    }
  }
  /**
   * Approves the currently pending plan.
   * Only valid if the session state is `awaitingPlanApproval`.
   *
   * **Side Effects:**
   * - Sends a POST request to `sessions/{id}:approvePlan`.
   * - Transitions the session state from `awaitingPlanApproval` to `inProgress` (eventually).
   *
   * @throws {InvalidStateError} If the session is not in the `awaitingPlanApproval` state.
   *
   * @example
   * await session.waitFor('awaitingPlanApproval');
   * await session.approve();
   */
  async approve() {
    await this.request(`sessions/${this.id}:approvePlan`, {
      method: "POST",
      body: {}
    });
  }
  /**
   * Sends a message (prompt) to the agent in the context of the current session.
   * This is a fire-and-forget operation. To see the response, use `stream()` or `ask()`.
   *
   * **Side Effects:**
   * - Sends a POST request to `sessions/{id}:sendMessage`.
   * - Appends a new `userMessaged` activity to the session history.
   *
   * @param prompt The message to send.
   *
   * @example
   * await session.send("Please clarify step 2.");
   */
  async send(prompt) {
    await this.request(`sessions/${this.id}:sendMessage`, {
      method: "POST",
      body: { prompt }
    });
  }
  /**
   * Sends a message to the agent and waits specifically for the agent's immediate reply.
   * This provides a convenient request/response flow for conversational interactions.
   *
   * **Behavior:**
   * - Sends the prompt using `send()`.
   * - Subscribes to the activity stream.
   * - Resolves with the first `agentMessaged` activity that appears *after* the prompt was sent.
   *
   * @param prompt The message to send.
   * @returns The agent's reply activity.
   * @throws {JulesError} If the session terminates before the agent replies.
   *
   * @example
   * const reply = await session.ask("What is the status?");
   * console.log(reply.message);
   */
  async ask(prompt) {
    const startTime = /* @__PURE__ */ new Date();
    await this.send(prompt);
    for await (const activity of this.stream({
      exclude: { originator: "user" }
    })) {
      const activityTime = new Date(activity.createTime).getTime();
      const askTime = startTime.getTime();
      if (activityTime <= askTime) {
        continue;
      }
      if (activity.type === "agentMessaged") {
        return activity;
      }
      if (activity.type === "sessionCompleted" || activity.type === "sessionFailed") {
        throw new JulesError("Session ended before the agent replied.");
      }
    }
    throw new JulesError("Session ended before the agent replied.");
  }
  /**
   * Waits for the session to reach a terminal state and returns the result.
   *
   * **Behavior:**
   * - Polls the session API until state is 'completed' or 'failed'.
   * - Maps the final session resource to a friendly `Outcome` object.
   *
   * @param options Optional configuration for the operation.
   * @param options.timeoutMs Maximum time in milliseconds to wait for the session to complete.
   * @returns The final outcome of the session.
   * @throws {AutomatedSessionFailedError} If the session ends in a 'failed' state.
   * @throws {TimeoutError} If the operation times out.
   */
  async result(options) {
    const finalSession = await pollUntilCompletion(
      this.id,
      this.apiClient,
      this.config.pollingIntervalMs,
      this.platform,
      options?.timeoutMs
    );
    await this.sessionStorage.upsert(finalSession);
    return mapSessionResourceToOutcome(finalSession);
  }
  /**
   * Pauses execution and waits until the session reaches a specific state.
   * Also returns if the session reaches a terminal state ('completed' or 'failed')
   * to prevent infinite waiting.
   *
   * **Behavior:**
   * - Polls the session API at the configured interval.
   * - Resolves immediately if the session is already in the target state (or terminal).
   *
   * @param targetState The target state to wait for.
   * @param options Optional configuration for the operation.
   * @param options.timeoutMs Maximum time in milliseconds to wait for the state.
   * @throws {TimeoutError} If the operation times out.
   *
   * @example
   * await session.waitFor('awaitingPlanApproval');
   */
  async waitFor(targetState, options) {
    await pollSession(
      this.id,
      this.apiClient,
      (session) => {
        const state = session.state;
        return state === targetState || state === "completed" || state === "failed";
      },
      this.config.pollingIntervalMs,
      this.platform,
      options?.timeoutMs
    );
  }
  /**
   * Archives the session.
   * This removes the session from the default list view and marks it as archived.
   * Archived sessions can still be accessed by ID or by filtering for `archived = true`.
   *
   * **Side Effects:**
   * - Sends a POST request to `sessions/{id}:archive`.
   * - Updates the local cache to mark the session as archived.
   */
  async archive() {
    await this.request(`sessions/${this.id}:archive`, {
      method: "POST",
      body: {}
    });
    const cached = await this.sessionStorage.get(this.id);
    if (cached) {
      const resource = { ...cached.resource, archived: true };
      await this.sessionStorage.upsert(resource);
    }
  }
  /**
   * Unarchives the session.
   * This restores the session to the default list view.
   *
   * **Side Effects:**
   * - Sends a POST request to `sessions/{id}:unarchive`.
   * - Updates the local cache to mark the session as not archived.
   */
  async unarchive() {
    await this.request(`sessions/${this.id}:unarchive`, {
      method: "POST",
      body: {}
    });
    const cached = await this.sessionStorage.get(this.id);
    if (cached) {
      const resource = { ...cached.resource, archived: false };
      await this.sessionStorage.upsert(resource);
    }
  }
  /**
   * Retrieves the latest state of the underlying session resource.
   * Implements "Iceberg" Read-Through caching.
   */
  async info() {
    let resource;
    const cached = await this.sessionStorage.get(this.id);
    if (isCacheValid(cached)) {
      resource = cached.resource;
    } else {
      try {
        const restResource = await this.request(
          `sessions/${this.id}`
        );
        resource = mapRestSessionToSdkSession(restResource, this.platform);
        await this.sessionStorage.upsert(resource);
      } catch (e) {
        if (e.status === 404 && cached) {
          await this.sessionStorage.delete(this.id);
        }
        throw e;
      }
    }
    resource.outcome = mapSessionResourceToOutcome(resource);
    return resource;
  }
  /**
   * Creates a point-in-time snapshot of the session.
   * This is a network operation that fetches the latest session info and all activities.
   *
   * @param options Optional configuration for the snapshot.
   * @param options.activities If true, includes all activities in the snapshot. Defaults to true.
   * @returns A `SessionSnapshot` instance.
   */
  async snapshot(options) {
    const includeActivities = options?.activities ?? true;
    const [info, activities] = await Promise.all([
      this.info(),
      includeActivities ? collectAsync(this.history()) : Promise.resolve([])
    ]);
    return new SessionSnapshotImpl({
      data: { session: info, activities: activities ?? [] }
    });
  }
}
async function pMap(items, mapper, options = {}) {
  const concurrency = options.concurrency ?? 3;
  const stopOnError = options.stopOnError ?? true;
  const delayMs = options.delayMs ?? 0;
  const results = new Array(items.length);
  const errors = new Array();
  let nextIndex = 0;
  const workers = new Array(concurrency).fill(0).map(async () => {
    while (true) {
      const index = nextIndex++;
      if (index >= items.length) {
        break;
      }
      const item = items[index];
      if (delayMs > 0) {
        await new Promise((resolve) => setTimeout(resolve, delayMs));
      }
      try {
        results[index] = await mapper(item, index);
      } catch (err) {
        if (stopOnError) {
          throw err;
        }
        errors.push(err);
      }
    }
  });
  await Promise.all(workers);
  if (!stopOnError && errors.length > 0) {
    throw new AggregateError(
      errors,
      "Multiple errors occurred during jules.all()"
    );
  }
  return results;
}
class SessionCursor {
  constructor(apiClient, storage, platform, options = {}) {
    this.apiClient = apiClient;
    this.storage = storage;
    this.platform = platform;
    this.options = options;
  }
  /**
   * DX Feature: Promise Compatibility.
   * Allows `const page = await jules.sessions()` to just get the first page.
   * This is great for UIs that render a list and a "Load More" button.
   */
  then(onfulfilled, onrejected) {
    return this.fetchPage(this.options.pageToken).then(onfulfilled, onrejected);
  }
  /**
   * DX Feature: Async Iterator.
   * Allows `for await (const s of jules.sessions())` to stream ALL items.
   * Automatically handles page tokens and fetching behind the scenes.
   */
  async *[Symbol.asyncIterator]() {
    let currentToken = this.options.pageToken;
    let itemCount = 0;
    const limit = this.options.limit ?? Infinity;
    do {
      if (itemCount >= limit) break;
      const response = await this.fetchPage(currentToken);
      if (!response.sessions || response.sessions.length === 0) {
        break;
      }
      for (const session of response.sessions) {
        if (itemCount >= limit) break;
        yield session;
        itemCount++;
      }
      currentToken = response.nextPageToken;
    } while (currentToken);
  }
  /**
   * Helper to fetch all pages into a single array.
   * WARNING: Use with caution on large datasets.
   */
  async all() {
    const results = [];
    for await (const session of this) {
      results.push(session);
    }
    return results;
  }
  /**
   * Internal fetcher that maps the options to the REST parameters.
   */
  async fetchPage(pageToken) {
    const params = {};
    if (this.options.pageSize)
      params.pageSize = this.options.pageSize.toString();
    if (pageToken) params.pageToken = pageToken;
    if (this.options.filter) params.filter = this.options.filter;
    const response = await this.apiClient.request("sessions", { query: params });
    const sessions = (response.sessions || []).map(
      (s) => mapRestSessionToSdkSession(s, this.platform)
    );
    if (sessions.length > 0 && this.options.persist !== false) {
      await this.storage.upsertMany(sessions);
    }
    return {
      sessions,
      nextPageToken: response.nextPageToken
    };
  }
}
const GLOBAL_METADATA_FILE = "global-metadata.json";
async function updateGlobalCacheMetadata(rootDirOverride) {
  const rootDir = getRootDir();
  const cacheDir = path$1.join(rootDir, ".jules/cache");
  const metadataPath = path$1.join(cacheDir, GLOBAL_METADATA_FILE);
  let metadata = { lastSyncedAt: 0, sessionCount: 0 };
  try {
    const content = await fs.readFile(metadataPath, "utf8");
    metadata = JSON.parse(content);
  } catch {
  }
  metadata.lastSyncedAt = Date.now();
  try {
    const entries = await fs.readdir(cacheDir, { withFileTypes: true });
    metadata.sessionCount = entries.filter((e) => e.isDirectory()).length;
  } catch {
    metadata.sessionCount = 0;
  }
  await fs.mkdir(cacheDir, { recursive: true });
  await fs.writeFile(metadataPath, JSON.stringify(metadata), "utf8");
}
function parseSelectExpression(expr) {
  if (expr === "*") {
    return { path: [], exclude: false, wildcard: true };
  }
  const exclude = expr.startsWith("-");
  const pathStr = exclude ? expr.slice(1) : expr;
  const cleanPath = pathStr.replace(/\[\]/g, "");
  const path2 = cleanPath.split(".").filter((p) => p.length > 0);
  return { path: path2, exclude, wildcard: false };
}
function getPath(obj, path2) {
  if (path2.length === 0) return obj;
  if (obj === null || obj === void 0) return void 0;
  const [head, ...tail] = path2;
  if (Array.isArray(obj)) {
    const results = obj.map((item) => getPath(item, path2)).filter((v) => v !== void 0);
    return results.length > 0 ? results : void 0;
  }
  if (typeof obj === "object") {
    const value = obj[head];
    return getPath(value, tail);
  }
  return void 0;
}
function setPath(obj, path2, value) {
  if (path2.length === 0 || value === void 0) return;
  const [head, ...tail] = path2;
  if (tail.length === 0) {
    obj[head] = value;
    return;
  }
  if (!(head in obj)) {
    obj[head] = {};
  }
  const next = obj[head];
  if (typeof next === "object" && next !== null && !Array.isArray(next)) {
    setPath(next, tail, value);
  }
}
function deletePath(obj, path2) {
  if (path2.length === 0 || obj === null || obj === void 0) return;
  if (Array.isArray(obj)) {
    obj.forEach((item) => deletePath(item, path2));
    return;
  }
  if (typeof obj !== "object") return;
  const record = obj;
  const [head, ...tail] = path2;
  if (tail.length === 0) {
    delete record[head];
    return;
  }
  if (head in record) {
    deletePath(record[head], tail);
  }
}
function deepClone(obj) {
  if (obj === null || typeof obj !== "object") return obj;
  if (Array.isArray(obj)) return obj.map((item) => deepClone(item));
  const cloned = {};
  for (const key of Object.keys(obj)) {
    cloned[key] = deepClone(obj[key]);
  }
  return cloned;
}
function projectArray(arr, subPaths, excludePaths) {
  return arr.map((item) => {
    if (item === null || typeof item !== "object") return item;
    const projected = {};
    for (const subPath of subPaths) {
      const value = getPath(item, subPath);
      if (value !== void 0) {
        if (subPath.length === 0) {
          Object.assign(projected, deepClone(item));
        } else {
          setPath(projected, subPath, deepClone(value));
        }
      }
    }
    for (const excludePath of excludePaths) {
      deletePath(projected, excludePath);
    }
    return projected;
  });
}
function projectDocument(doc, selects) {
  if (!selects || selects.length === 0) {
    return doc;
  }
  const parsed = selects.map(parseSelectExpression);
  const hasWildcard = parsed.some((p) => p.wildcard && !p.exclude);
  const inclusions = parsed.filter((p) => !p.exclude && !p.wildcard);
  const exclusions = parsed.filter((p) => p.exclude);
  let result;
  if (hasWildcard) {
    result = deepClone(doc);
  } else {
    result = {};
    const byTopLevel = /* @__PURE__ */ new Map();
    for (const incl of inclusions) {
      if (incl.path.length === 0) continue;
      const top = incl.path[0];
      if (!byTopLevel.has(top)) {
        byTopLevel.set(top, []);
      }
      byTopLevel.get(top).push(incl.path.slice(1));
    }
    for (const [topField, subPaths] of byTopLevel) {
      const value = doc[topField];
      if (value === void 0) continue;
      if (Array.isArray(value)) {
        const exclusionSubPaths = exclusions.filter((e) => e.path[0] === topField).map((e) => e.path.slice(1));
        if (subPaths.some((p) => p.length === 0)) {
          result[topField] = projectArray(value, [[]], exclusionSubPaths);
        } else {
          result[topField] = projectArray(value, subPaths, exclusionSubPaths);
        }
      } else if (typeof value === "object" && value !== null) {
        if (subPaths.some((p) => p.length === 0)) {
          result[topField] = deepClone(value);
        } else {
          const nestedSelects = subPaths.map((p) => p.join("."));
          result[topField] = projectDocument(
            value,
            nestedSelects
          );
        }
      } else {
        result[topField] = value;
      }
    }
  }
  for (const excl of exclusions) {
    deletePath(result, excl.path);
  }
  return result;
}
const MAX_SUMMARY_LENGTH = 200;
function toSummary(activity) {
  const { id, type, createTime } = activity;
  let summary = type;
  switch (activity.type) {
    case "agentMessaged":
    case "userMessaged": {
      const message = activity.message;
      if (!message || message.length === 0) {
        summary = type;
      } else if (message.length > MAX_SUMMARY_LENGTH) {
        summary = message.substring(0, MAX_SUMMARY_LENGTH) + "...";
      } else {
        summary = message;
      }
      break;
    }
    case "progressUpdated": {
      const progress = activity;
      if (progress.title && progress.description) {
        summary = `${progress.title}: ${progress.description}`;
      } else if (progress.title) {
        summary = progress.title;
      } else if (progress.description) {
        summary = progress.description;
      }
      break;
    }
    case "planGenerated": {
      const plan = activity;
      const stepCount = plan.plan?.steps?.length ?? 0;
      summary = `Plan generated with ${stepCount} steps`;
      break;
    }
    case "planApproved":
      summary = "Plan approved";
      break;
    case "sessionCompleted":
      summary = "Session completed";
      break;
    case "sessionFailed": {
      const failed = activity;
      summary = failed.reason ? `Session failed: ${failed.reason}` : "Session failed";
      break;
    }
  }
  return { id, type, createTime, summary };
}
function computeArtifactCount(activity) {
  return activity.artifacts?.length ?? 0;
}
function computeSummary(activity) {
  return toSummary(activity).summary;
}
function computeDurationMs(session) {
  if (!session.createTime || !session.updateTime) return 0;
  const created = new Date(session.createTime).getTime();
  const updated = new Date(session.updateTime).getTime();
  if (isNaN(created) || isNaN(updated)) return 0;
  return Math.max(0, updated - created);
}
function injectActivityComputedFields(activity, selectFields) {
  const result = { ...activity };
  const includeAll = !selectFields || selectFields.length === 0 || selectFields.includes("*");
  const needsArtifactCount = includeAll || selectFields?.includes("artifactCount");
  const needsSummary = includeAll || selectFields?.includes("summary");
  if (needsArtifactCount) {
    result.artifactCount = computeArtifactCount(activity);
  }
  if (needsSummary) {
    result.summary = computeSummary(activity);
  }
  return result;
}
function injectSessionComputedFields(session, selectFields) {
  const result = { ...session };
  const includeAll = !selectFields || selectFields.length === 0 || selectFields.includes("*");
  const needsDurationMs = includeAll || selectFields?.includes("durationMs");
  if (needsDurationMs) {
    result.durationMs = computeDurationMs(session);
  }
  return result;
}
const DEFAULT_ACTIVITY_PROJECTION = [
  "id",
  "type",
  "createTime",
  "originator",
  "artifactCount",
  "summary"
];
const DEFAULT_SESSION_PROJECTION = [
  "id",
  "state",
  "title",
  "createTime"
];
function match(actual, filter) {
  if (filter === void 0) return true;
  if (typeof filter !== "object" || filter === null || Array.isArray(filter)) {
    return actual === filter;
  }
  const op = filter;
  if (op.exists !== void 0) {
    const valueExists = actual !== void 0 && actual !== null;
    return op.exists ? valueExists : !valueExists;
  }
  if (op.eq !== void 0 && actual !== op.eq) return false;
  if (op.neq !== void 0 && actual === op.neq) return false;
  if (op.contains !== void 0 && typeof actual === "string" && !actual.toLowerCase().includes(op.contains.toLowerCase()))
    return false;
  if (op.gt !== void 0 && op.gt !== null && actual <= op.gt) return false;
  if (op.gte !== void 0 && op.gte !== null && actual < op.gte) return false;
  if (op.lt !== void 0 && op.lt !== null && actual >= op.lt) return false;
  if (op.lte !== void 0 && op.lte !== null && actual > op.lte) return false;
  if (op.in !== void 0 && !op.in.includes(actual)) return false;
  return true;
}
function isDotPath(key) {
  return key.includes(".");
}
function matchPath(doc, path2, filter) {
  const pathParts = path2.split(".");
  const value = getPath(doc, pathParts);
  if (Array.isArray(value)) {
    return value.some((v) => match(v, filter));
  }
  return match(value, filter);
}
function matchWhere(doc, where) {
  if (!where) return true;
  for (const [key, filter] of Object.entries(where)) {
    if (isDotPath(key)) {
      if (!matchPath(doc, key, filter)) return false;
    } else {
      const value = doc[key];
      if (!match(value, filter)) return false;
    }
  }
  return true;
}
function toActivitySelectOptions(where) {
  if (!where) return {};
  const options = {};
  if (where.type) {
    if (typeof where.type === "string") {
      options.type = where.type;
    } else if (typeof where.type === "object" && "eq" in where.type && where.type.eq) {
      options.type = where.type.eq;
    }
  }
  return options;
}
function applyProjection(doc, select2, domain) {
  const docRecord = doc;
  const withComputed = domain === "activities" ? injectActivityComputedFields(doc, select2) : injectSessionComputedFields(docRecord, select2);
  if (!select2) {
    const defaults = domain === "activities" ? DEFAULT_ACTIVITY_PROJECTION : DEFAULT_SESSION_PROJECTION;
    return projectDocument(withComputed, defaults);
  }
  if (select2.length === 0) {
    return withComputed;
  }
  return projectDocument(withComputed, select2);
}
async function select(client, query) {
  const storage = client.storage;
  const results = [];
  const limit = query.limit ?? Infinity;
  if (query.from === "sessions") {
    const where = query.where;
    const whereRecord = where;
    const dotFilters = whereRecord ? Object.entries(whereRecord).filter(([k]) => isDotPath(k)) : [];
    const dotWhere = dotFilters.length > 0 ? Object.fromEntries(dotFilters) : void 0;
    let chunk = [];
    const CHUNK_SIZE = 50;
    const processChunk = async () => {
      if (chunk.length === 0) return;
      const hydrated = await pMap(
        chunk,
        async (entry) => {
          const cached = await storage.get(entry.id);
          return { entry, cached };
        },
        { concurrency: 10 }
      );
      for (const { cached } of hydrated) {
        if (results.length >= limit) break;
        if (!cached) continue;
        if (dotWhere && !matchWhere(cached.resource, dotWhere)) continue;
        const item = applyProjection(
          cached.resource,
          query.select,
          "sessions"
        );
        const resourceRecord = cached.resource;
        item._sortKey = {
          createTime: resourceRecord.createTime,
          id: resourceRecord.id
        };
        results.push(item);
      }
      chunk = [];
    };
    for await (const entry of storage.scanIndex()) {
      if (results.length >= limit) break;
      if (where?.id && !match(entry.id, where.id)) continue;
      if (where?.state && !match(entry.state, where.state)) continue;
      if (where?.title && !match(entry.title, where.title)) continue;
      if (where?.search && !entry.title.toLowerCase().includes(where.search.toLowerCase()))
        continue;
      chunk.push(entry);
      if (chunk.length >= CHUNK_SIZE || !dotWhere && chunk.length >= limit - results.length) {
        await processChunk();
      }
    }
    await processChunk();
    if (query.include && "activities" in query.include) {
      const actConfig = query.include.activities;
      let mappedOptions = {};
      if (typeof actConfig === "object") {
        mappedOptions = {
          ...toActivitySelectOptions(actConfig.where),
          limit: actConfig.limit
        };
      }
      await pMap(
        results,
        async (session) => {
          const sessionClient = await client.session(session.id);
          const localActivities = await sessionClient.activities.select({});
          const activities = [];
          for (const act of localActivities) {
            if (mappedOptions.limit && activities.length >= mappedOptions.limit) {
              break;
            }
            if (mappedOptions.type && act.type !== mappedOptions.type) {
              continue;
            }
            activities.push(act);
          }
          session.activities = activities;
        },
        { concurrency: 5 }
      );
    }
  } else if (query.from === "activities") {
    const where = query.where;
    let targetSessionIds = [];
    if (where?.sessionId) {
      if (typeof where.sessionId === "string") {
        targetSessionIds = [where.sessionId];
      } else if (typeof where.sessionId === "object" && "eq" in where.sessionId && where.sessionId.eq) {
        targetSessionIds = [where.sessionId.eq];
      }
    }
    const sessionCache = /* @__PURE__ */ new Map();
    const sessionScanner = async function* () {
      if (targetSessionIds.length > 0) {
        for (const id of targetSessionIds) {
          yield { id };
        }
      } else {
        yield* storage.scanIndex();
      }
    };
    const sessionEntries = [];
    for await (const sessionEntry of sessionScanner()) {
      sessionEntries.push(sessionEntry);
    }
    const sessionResults = await pMap(
      sessionEntries,
      async (sessionEntry) => {
        const sessionClient = await client.session(sessionEntry.id);
        const localActivities = await sessionClient.activities.select({});
        const filtered = [];
        for (const act of localActivities) {
          if (where?.id && !match(act.id, where.id)) continue;
          if (where?.type && !match(act.type, where.type)) continue;
          const activityWhere = where ? Object.fromEntries(
            Object.entries(where).filter(([k]) => k !== "sessionId")
          ) : void 0;
          if (!matchWhere(act, activityWhere)) continue;
          const item = applyProjection(
            act,
            query.select,
            "activities"
          );
          const actRecord = act;
          item._sortKey = {
            createTime: actRecord.createTime,
            id: actRecord.id
          };
          if (query.include && "session" in query.include) {
            const sessConfig = query.include.session;
            const sessSelect = typeof sessConfig === "object" ? sessConfig.select : void 0;
            let sessionInfo = sessionCache.get(sessionEntry.id);
            if (!sessionInfo) {
              const info = await sessionClient.info();
              sessionInfo = info;
              sessionCache.set(sessionEntry.id, sessionInfo);
            }
            item.session = applyProjection(
              sessionInfo,
              sessSelect,
              "sessions"
            );
          }
          filtered.push(item);
        }
        return filtered;
      },
      { concurrency: 5 }
    );
    for (const res of sessionResults) {
      results.push(...res);
    }
  }
  const order = query.order ?? "desc";
  results.sort((a, b) => {
    const sortKeyA = a._sortKey;
    const sortKeyB = b._sortKey;
    const timeA = new Date(
      sortKeyA?.createTime ?? a.createTime
    ).getTime();
    const timeB = new Date(
      sortKeyB?.createTime ?? b.createTime
    ).getTime();
    const idA = sortKeyA?.id ?? a.id;
    const idB = sortKeyB?.id ?? b.id;
    if (timeA !== timeB) {
      return order === "desc" ? timeB - timeA : timeA - timeB;
    }
    if (order === "desc") {
      return idB.localeCompare(idA);
    }
    return idA.localeCompare(idB);
  });
  let finalResults = results;
  const cursorId = query.startAfter ?? query.startAt;
  if (cursorId) {
    const cursorIndex = finalResults.findIndex((item) => {
      const sortKey = item._sortKey;
      const itemId = sortKey?.id ?? item.id;
      return itemId === cursorId;
    });
    if (cursorIndex === -1) {
      return [];
    }
    const sliceIndex = query.startAfter ? cursorIndex + 1 : cursorIndex;
    finalResults = finalResults.slice(sliceIndex);
  }
  for (const result of finalResults) {
    delete result._sortKey;
  }
  return finalResults.slice(0, limit);
}
class JulesClientImpl {
  /**
   * Manages source connections (e.g., GitHub repositories).
   */
  sources;
  // Expose storage for modular functions (Phase 3 requirement)
  storage;
  apiClient;
  config;
  options;
  storageFactory;
  platform;
  /**
   * Lock to prevent concurrent sync operations.
   * Using a simple boolean for in-process locking.
   */
  syncInProgress = false;
  /**
   * Creates a new instance of the JulesClient.
   *
   * @param options Configuration options for the client.
   * @param defaultStorageFactory Factory for creating storage instances.
   * @param defaultPlatform Platform-specific implementation.
   */
  constructor(options = {}, defaultStorageFactory2, defaultPlatform2) {
    this.options = options;
    this.storageFactory = options.storageFactory ?? defaultStorageFactory2;
    this.platform = options.platform ?? defaultPlatform2;
    this.storage = this.storageFactory.session();
    const apiKey = options.apiKey_TEST_ONLY_DO_NOT_USE_IN_PRODUCTION ?? options.apiKey ?? this.platform.getEnv("JULES_API_KEY");
    const baseUrl = options.baseUrl ?? "https://jules.googleapis.com/v1alpha";
    this.config = {
      pollingIntervalMs: options.config?.pollingIntervalMs ?? 5e3,
      requestTimeoutMs: options.config?.requestTimeoutMs ?? 3e4
    };
    this.apiClient = new ApiClient({
      apiKey,
      baseUrl,
      requestTimeoutMs: this.config.requestTimeoutMs,
      rateLimitRetry: options.config?.rateLimitRetry
    });
    this.sources = createSourceManager(this.apiClient);
  }
  /**
   * Fluent API for rich local querying across sessions and activities.
   * This method uses the modular query engine internally.
   */
  async select(query) {
    return select(this, query);
  }
  /**
   * Synchronizes local state with the server.
   * Logic:
   * 1. Find High-Water Mark (newest local record).
   * 2. Stream latest sessions from API.
   * 3. Terminate stream early if 'incremental' and High-Water Mark is hit.
   * 4. Throttled hydration of activities if depth is 'activities'.
   */
  async sync(options = {}) {
    if (this.syncInProgress) {
      throw new SyncInProgressError();
    }
    this.syncInProgress = true;
    try {
      const startTime = Date.now();
      const {
        sessionId,
        limit = 100,
        depth = "metadata",
        incremental = true,
        concurrency = 3,
        onProgress,
        checkpoint: useCheckpoint = false,
        signal
      } = options;
      let wasAborted = false;
      const candidates = [];
      let activitiesIngested = 0;
      let sessionsIngestedThisRun = 0;
      if (sessionId) {
        const restSession = await this.apiClient.request(
          `sessions/${sessionId}`
        );
        const session = mapRestSessionToSdkSession(restSession, this.platform);
        await this.storage.upsert(session);
        candidates.push(session);
        sessionsIngestedThisRun = 1;
      } else {
        let resumeFromId = null;
        let startingCount = 0;
        if (useCheckpoint) {
          const ckpt = await this.loadCheckpoint();
          if (ckpt) {
            resumeFromId = ckpt.lastProcessedSessionId;
            startingCount = ckpt.sessionsProcessed;
          }
        }
        let skipUntilPast = !!resumeFromId;
        const highWaterMark = incremental ? await this._getHighWaterMark() : null;
        const cursor = this.sessions({
          pageSize: Math.min(limit, 100),
          persist: false
        });
        onProgress?.({ phase: "fetching_list", current: 0 });
        for await (const session of cursor) {
          if (signal?.aborted) {
            wasAborted = true;
            break;
          }
          if (skipUntilPast) {
            if (session.id === resumeFromId) {
              skipUntilPast = false;
              continue;
            }
            continue;
          }
          if (highWaterMark && new Date(session.createTime) <= highWaterMark) {
            if (depth === "activities") {
              await this.storage.upsert(session);
              candidates.push(session);
            }
            break;
          }
          await this.storage.upsert(session);
          candidates.push(session);
          sessionsIngestedThisRun++;
          if (useCheckpoint) {
            await this.saveCheckpoint({
              lastProcessedSessionId: session.id,
              sessionsProcessed: startingCount + sessionsIngestedThisRun,
              startedAt: new Date(startTime).toISOString()
            });
          }
          onProgress?.({
            phase: "fetching_list",
            current: sessionsIngestedThisRun,
            lastIngestedId: session.id
          });
          if (candidates.length >= limit) break;
        }
      }
      if (depth === "activities" && candidates.length > 0 && !wasAborted) {
        let hydratedCount = 0;
        onProgress?.({
          phase: "hydrating_records",
          current: 0,
          total: candidates.length
        });
        await pMap(
          candidates,
          async (session) => {
            if (signal?.aborted) return;
            const sessionClient = this.session(session.id);
            const count = await sessionClient.activities.hydrate();
            activitiesIngested += count;
            hydratedCount++;
            onProgress?.({
              phase: "hydrating_records",
              current: hydratedCount,
              total: candidates.length,
              lastIngestedId: session.id,
              activityCount: count
            });
          },
          { concurrency }
        );
      }
      if (useCheckpoint && !wasAborted && !sessionId) {
        await this.clearCheckpoint();
      }
      const stats = {
        sessionsIngested: sessionsIngestedThisRun,
        activitiesIngested,
        isComplete: !wasAborted,
        // false if aborted
        durationMs: Date.now() - startTime
      };
      await updateGlobalCacheMetadata();
      return stats;
    } finally {
      this.syncInProgress = false;
    }
  }
  getCheckpointPath() {
    return join(getRootDir(), ".jules", "cache", "sync-checkpoint.json");
  }
  async loadCheckpoint() {
    if (!this.platform.readFile) return null;
    try {
      const path2 = this.getCheckpointPath();
      const data = await this.platform.readFile(path2);
      return JSON.parse(data);
    } catch {
      return null;
    }
  }
  async saveCheckpoint(checkpoint) {
    if (!this.platform.writeFile) return;
    const path2 = this.getCheckpointPath();
    await this.platform.writeFile(path2, JSON.stringify(checkpoint, null, 2));
  }
  async clearCheckpoint() {
    if (!this.platform.deleteFile) return;
    try {
      const path2 = this.getCheckpointPath();
      await this.platform.deleteFile(path2);
    } catch {
    }
  }
  async _getHighWaterMark() {
    let newest = null;
    for await (const entry of this.storage.scanIndex()) {
      const date = new Date(entry.createTime);
      if (!newest || date > newest) newest = date;
    }
    return newest;
  }
  /**
   * Helper to resolve environment variables with support for frontend prefixes.
   */
  getEnv(key) {
    return this.platform.getEnv(`NEXT_PUBLIC_${key}`) || this.platform.getEnv(`REACT_APP_${key}`) || this.platform.getEnv(`VITE_${key}`) || this.platform.getEnv(key);
  }
  /**
   * Creates a new Jules client instance with updated configuration.
   * This is an immutable operation; the original client instance remains unchanged.
   *
   * @param options The new configuration options to merge with the existing ones.
   * @returns A new JulesClient instance with the updated configuration.
   */
  with(options) {
    return new JulesClientImpl(
      {
        ...this.options,
        ...options,
        config: {
          ...this.options.config,
          ...options.config
        }
      },
      this.storageFactory,
      this.platform
    );
  }
  /**
   * Connects to the Jules service with the provided configuration.
   * Acts as a factory method for creating a new client instance.
   *
   * @param options Configuration options for the client.
   * @returns A new JulesClient instance.
   */
  connect(options) {
    return new JulesClientImpl(
      {
        ...this.options,
        ...options
      },
      this.storageFactory,
      this.platform
    );
  }
  /**
   * Retrieves a session resource using the "Iceberg" caching strategy.
   * * - **Tier 3 (Frozen):** > 30 days old. Returns from cache immediately.
   * - **Tier 2 (Warm):** Terminal state + Verified < 24h ago. Returns from cache.
   * - **Tier 1 (Hot):** Active or Stale. Fetches from network, updates cache, returns.
   */
  async getSessionResource(id) {
    const cached = await this.storage.get(id);
    if (isCacheValid(cached)) {
      return cached.resource;
    }
    try {
      const restFresh = await this.apiClient.request(
        `sessions/${id}`
      );
      const fresh = mapRestSessionToSdkSession(restFresh, this.platform);
      await this.storage.upsert(fresh);
      return fresh;
    } catch (e) {
      if (e.status === 404 && cached) {
        await this.storage.delete(id);
      }
      throw e;
    }
  }
  /**
   * Lists sessions with a fluent, pagination-friendly API.
   * @param options Configuration for pagination (pageSize, limit, pageToken)
   * @returns A SessionCursor that can be awaited (first page) or iterated (all pages).
   */
  sessions(options) {
    return new SessionCursor(
      this.apiClient,
      this.storage,
      this.platform,
      options
    );
  }
  async all(items, mapper, options) {
    return pMap(
      items,
      async (item) => {
        const config = await mapper(item);
        return this.run(config);
      },
      options
    );
  }
  async _prepareSessionCreation(config) {
    if (!config.source) {
      return {
        prompt: config.prompt,
        title: config.title
      };
    }
    const source = await this.sources.get({ github: config.source.github });
    if (!source) {
      throw new SourceNotFoundError(config.source.github);
    }
    return {
      prompt: config.prompt,
      title: config.title,
      sourceContext: {
        source: source.name,
        githubRepoContext: {
          startingBranch: config.source.baseBranch
        }
      }
    };
  }
  /**
   * Executes a task in automated mode.
   * This is a high-level abstraction for "fire-and-forget" tasks.
   *
   * **Side Effects:**
   * - Creates a new session on the Jules API (`POST /sessions`).
   * - Initiates background polling for activity updates.
   * - May create a Pull Request if `autoPr` is true (default).
   *
   * **Data Transformation:**
   * - Resolves the `github` source identifier (e.g., `owner/repo`) to a full resource name.
   * - Defaults `requirePlanApproval` to `false` for automated runs.
   *
   * @param config The configuration for the run.
   * @returns A `AutomatedSession` object, which is an enhanced Promise that resolves to the final outcome.
   * @throws {SourceNotFoundError} If the specified GitHub repository cannot be found or accessed.
   * @throws {JulesApiError} If the session creation fails (e.g., 401 Unauthorized).
   *
   * @example
   * const run = await jules.run({
   *   prompt: "Fix the login bug",
   *   source: { github: "my-org/repo", baseBranch: "main" }
   * });
   * const outcome = await run.result();
   */
  async run(config) {
    const body = await this._prepareSessionCreation(config);
    const restSessionResource = await this.apiClient.request("sessions", {
      method: "POST",
      body: {
        ...body,
        automationMode: config.autoPr === false ? "AUTOMATION_MODE_UNSPECIFIED" : "AUTO_CREATE_PR",
        requirePlanApproval: config.requireApproval ?? false
      }
    });
    const sessionResource = mapRestSessionToSdkSession(
      restSessionResource,
      this.platform
    );
    await this.storage.upsert(sessionResource);
    const sessionId = sessionResource.id;
    return {
      id: sessionId,
      stream: (async function* () {
        yield* streamActivities(
          sessionId,
          this.apiClient,
          this.config.pollingIntervalMs,
          this.platform
        );
      }).bind(this),
      result: async () => {
        const finalSession = await pollUntilCompletion(
          sessionId,
          this.apiClient,
          this.config.pollingIntervalMs,
          this.platform
        );
        await this.storage.upsert(finalSession);
        return mapSessionResourceToOutcome(finalSession);
      }
    };
  }
  session(configOrId) {
    if (typeof configOrId === "string") {
      const storage = this.storageFactory.activity(configOrId);
      return new SessionClientImpl(
        configOrId,
        this.apiClient,
        this.config,
        storage,
        this.storage,
        this.platform
      );
    }
    const config = configOrId;
    const sessionPromise = (async () => {
      const body = await this._prepareSessionCreation(config);
      const restSession = await this.apiClient.request(
        "sessions",
        {
          method: "POST",
          body: {
            ...body,
            automationMode: config.autoPr === false ? "AUTOMATION_MODE_UNSPECIFIED" : "AUTO_CREATE_PR",
            requirePlanApproval: config.requireApproval ?? true
          }
        }
      );
      const session = mapRestSessionToSdkSession(restSession, this.platform);
      await this.storage.upsert(session);
      const activityStorage = this.storageFactory.activity(session.id);
      return new SessionClientImpl(
        session.name,
        this.apiClient,
        this.config,
        activityStorage,
        this.storage,
        this.platform
      );
    })();
    return sessionPromise;
  }
}
class NodeFileStorage {
  filePath;
  metadataPath;
  initialized = false;
  writeStream = null;
  // In-memory index: ActivityID -> Byte Offset
  index = /* @__PURE__ */ new Map();
  indexBuilt = false;
  indexBuildPromise = null;
  // Tracks the current file size to calculate offsets for new appends
  currentFileSize = 0;
  constructor(sessionId, rootDir) {
    const sessionCacheDir = path$1.resolve(rootDir, ".jules/cache", sessionId);
    this.filePath = path$1.join(sessionCacheDir, "activities.jsonl");
    this.metadataPath = path$1.join(sessionCacheDir, "metadata.json");
  }
  /**
   * Initializes the storage by ensuring the cache directory exists.
   *
   * **Side Effects:**
   * - Creates the `.jules/cache/<sessionId>` directory if it does not exist.
   * - Sets the internal `initialized` flag.
   */
  async init() {
    if (this.initialized) return;
    await fs.mkdir(path$1.dirname(this.filePath), { recursive: true });
    try {
      const stats = await fs.stat(this.filePath);
      this.currentFileSize = stats.size;
    } catch (e) {
      if (e.code === "ENOENT") {
        this.currentFileSize = 0;
      } else {
        throw e;
      }
    }
    this.writeStream = createWriteStream(this.filePath, {
      flags: "a",
      encoding: "utf8"
    });
    this.writeStream.on("error", (err) => {
      console.error(
        `[NodeFileStorage] WriteStream error for ${this.filePath}:`,
        err
      );
    });
    this.initialized = true;
  }
  /**
   * Closes the storage.
   */
  async close() {
    if (this.writeStream) {
      await new Promise((resolve) => this.writeStream.end(resolve));
      this.writeStream = null;
    }
    this.initialized = false;
    this.indexBuilt = false;
    this.index.clear();
    this.indexBuildPromise = null;
  }
  async _readMetadata() {
    try {
      const content = await fs.readFile(this.metadataPath, "utf8");
      return JSON.parse(content);
    } catch (e) {
      if (e.code === "ENOENT") {
        return { activityCount: 0 };
      }
      throw e;
    }
  }
  async _writeMetadata(metadata) {
    await fs.writeFile(
      this.metadataPath,
      JSON.stringify(metadata, null, 2),
      "utf8"
    );
  }
  /**
   * Appends an activity to the file.
   *
   * **Side Effects:**
   * - Appends a new line containing the JSON representation of the activity to `activities.jsonl`.
   * - Implicitly calls `init()` if not already initialized.
   */
  async append(activity) {
    if (!this.initialized) await this.init();
    const metadata = await this._readMetadata();
    metadata.activityCount += 1;
    await this._writeMetadata(metadata);
    const line = JSON.stringify(activity) + "\n";
    const startOffset = this.currentFileSize;
    if (this.writeStream) {
      const canContinue = this.writeStream.write(line);
      this.currentFileSize += Buffer.byteLength(line);
      if (this.indexBuilt || this.indexBuildPromise) {
        if (!this.index.has(activity.id)) {
          this.index.set(activity.id, startOffset);
        }
      }
      if (!canContinue) {
        await new Promise(
          (resolve) => this.writeStream.once("drain", resolve)
        );
      }
    } else {
      throw new Error("NodeFileStorage: WriteStream is not initialized");
    }
  }
  /**
   * Builds the in-memory index by scanning the file once.
   * Handles concurrency by coalescing multiple calls into a single promise.
   */
  async buildIndex() {
    if (this.indexBuilt) return;
    if (this.indexBuildPromise) return this.indexBuildPromise;
    this.indexBuildPromise = (async () => {
      try {
        this.index.clear();
        try {
          await fs.access(this.filePath);
        } catch (e) {
          this.indexBuilt = true;
          return;
        }
        const fileStream = createReadStream(this.filePath, {
          encoding: "utf8"
        });
        const rl = readline.createInterface({
          input: fileStream,
          crlfDelay: Infinity
        });
        let currentOffset = 0;
        for await (const line of rl) {
          const byteLen = Buffer.byteLength(line);
          const lineTotalBytes = byteLen + 1;
          if (line.trim().length > 0) {
            try {
              const activity = JSON.parse(line);
              if (!this.index.has(activity.id)) {
                this.index.set(activity.id, currentOffset);
              }
            } catch (e) {
            }
          }
          currentOffset += lineTotalBytes;
        }
        this.indexBuilt = true;
      } finally {
        this.indexBuildPromise = null;
      }
    })();
    return this.indexBuildPromise;
  }
  /**
   * Retrieves an activity by ID.
   * Uses an in-memory index (ID -> Offset) to seek directly to the line.
   */
  async get(activityId) {
    if (!this.initialized) await this.init();
    if (!this.indexBuilt) await this.buildIndex();
    const offset = this.index.get(activityId);
    if (offset === void 0) return void 0;
    return new Promise((resolve, reject) => {
      const stream = createReadStream(this.filePath, {
        start: offset,
        encoding: "utf8"
      });
      const rl = readline.createInterface({
        input: stream,
        crlfDelay: Infinity
      });
      let found = false;
      rl.on("line", (line) => {
        if (found) return;
        found = true;
        rl.close();
        stream.destroy();
        try {
          const activity = JSON.parse(line);
          resolve(activity);
        } catch (e) {
          resolve(void 0);
        }
      });
      rl.on("error", (err) => {
        reject(err);
      });
      stream.on("error", (err) => {
        reject(err);
      });
      rl.on("close", () => {
        if (!found) resolve(void 0);
      });
    });
  }
  /**
   * Retrieves the latest activity.
   * Efficiently reads the file backwards to find the last valid entry.
   */
  async latest() {
    if (!this.initialized) await this.init();
    try {
      await fs.access(this.filePath);
    } catch (e) {
      if (e.code === "ENOENT") {
        return void 0;
      }
      throw e;
    }
    const stat = await fs.stat(this.filePath);
    const fileSize = stat.size;
    if (fileSize === 0) return void 0;
    const bufferSize = 4096;
    const buffer = Buffer.alloc(bufferSize);
    let fd;
    try {
      fd = await fs.open(this.filePath, "r");
      let currentPos = fileSize;
      let trailing = "";
      while (currentPos > 0) {
        const readSize = Math.min(bufferSize, currentPos);
        const position = currentPos - readSize;
        const result = await fd.read(buffer, 0, readSize, position);
        const chunk = result.buffer.toString("utf8", 0, readSize);
        const content = chunk + trailing;
        const lines = content.split("\n");
        if (position > 0) {
          trailing = lines.shift() || "";
        } else {
          trailing = "";
        }
        for (let i = lines.length - 1; i >= 0; i--) {
          const line = lines[i].trim();
          if (line.length === 0) continue;
          try {
            return JSON.parse(line);
          } catch (e) {
            console.warn(
              `[NodeFileStorage] Corrupt JSON line ignored during latest() check in ${this.filePath}`
            );
          }
        }
        currentPos -= readSize;
      }
    } finally {
      if (fd) await fd.close();
    }
    return void 0;
  }
  /**
   * Yields all activities in the file.
   *
   * **Behavior:**
   * - Opens a read stream to `activities.jsonl`.
   * - Reads line-by-line using `readline`.
   * - Parses each line as JSON.
   *
   * **Edge Cases:**
   * - Logs a warning and skips lines if JSON parsing fails (corrupt data).
   * - Returns immediately (yields nothing) if the file does not exist.
   */
  async *scan() {
    if (!this.initialized) await this.init();
    try {
      await fs.access(this.filePath);
    } catch (e) {
      if (e.code === "ENOENT") {
        return;
      }
      throw e;
    }
    const fileStream = createReadStream(this.filePath, { encoding: "utf8" });
    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity
    });
    for await (const line of rl) {
      if (line.trim().length === 0) continue;
      try {
        yield JSON.parse(line);
      } catch (e) {
        console.warn(
          `[NodeFileStorage] Corrupt JSON line ignored in ${this.filePath}`
        );
      }
    }
  }
}
class NodeSessionStorage {
  cacheDir;
  indexFilePath;
  initialized = false;
  constructor(rootDir) {
    this.cacheDir = path$1.resolve(rootDir, ".jules/cache");
    this.indexFilePath = path$1.join(this.cacheDir, "sessions.jsonl");
  }
  async init() {
    if (this.initialized) return;
    await fs.mkdir(this.cacheDir, { recursive: true });
    this.initialized = true;
  }
  getSessionPath(sessionId) {
    return path$1.join(this.cacheDir, sessionId, "session.json");
  }
  async upsert(session) {
    await this.init();
    const sessionDir = path$1.join(this.cacheDir, session.id);
    await fs.mkdir(sessionDir, { recursive: true });
    const cached = {
      resource: session,
      _lastSyncedAt: Date.now()
    };
    await fs.writeFile(
      path$1.join(sessionDir, "session.json"),
      JSON.stringify(cached, null, 2),
      "utf8"
    );
    const indexEntry = {
      id: session.id,
      title: session.title,
      state: session.state,
      createTime: session.createTime,
      source: session.sourceContext?.source || "unknown",
      _updatedAt: Date.now()
    };
    await fs.appendFile(
      this.indexFilePath,
      JSON.stringify(indexEntry) + "\n",
      "utf8"
    );
  }
  async upsertMany(sessions) {
    await Promise.all(sessions.map((s) => this.upsert(s)));
  }
  async get(sessionId) {
    await this.init();
    try {
      const data = await fs.readFile(this.getSessionPath(sessionId), "utf8");
      return JSON.parse(data);
    } catch (e) {
      if (e.code === "ENOENT") return void 0;
      throw e;
    }
  }
  async delete(sessionId) {
    await this.init();
    const sessionDir = path$1.join(this.cacheDir, sessionId);
    await fs.rm(sessionDir, { recursive: true, force: true });
  }
  async *scanIndex() {
    await this.init();
    try {
      const fileStream = createReadStream(this.indexFilePath, {
        encoding: "utf8"
      });
      const rl = readline.createInterface({
        input: fileStream,
        crlfDelay: Infinity
      });
      const entries = /* @__PURE__ */ new Map();
      for await (const line of rl) {
        if (!line.trim()) continue;
        try {
          const entry = JSON.parse(line);
          entries.set(entry.id, entry);
        } catch (e) {
        }
      }
      for (const entry of entries.values()) {
        yield entry;
      }
    } catch (e) {
      if (e.code === "ENOENT") return;
      throw e;
    }
  }
}
class NodePlatform {
  /**
   * Saves a file to the local filesystem using `node:fs/promises`.
   *
   * **Side Effects:**
   * - Writes a file to disk.
   * - Overwrites the file if it already exists.
   */
  async saveFile(filepath, data, encoding, activityId) {
    const buffer = Buffer$1.from(data, encoding);
    await writeFile(filepath, buffer);
  }
  async sleep(ms) {
    await setTimeout$1(ms);
  }
  createDataUrl(data, mimeType) {
    return `data:${mimeType};base64,${data}`;
  }
  async fetch(input, init) {
    const res = await global.fetch(input, init);
    return {
      ok: res.ok,
      status: res.status,
      json: () => res.json(),
      text: () => res.text()
    };
  }
  crypto = {
    randomUUID: () => crypto.randomUUID(),
    async sign(text, secret) {
      const hmac = crypto.createHmac("sha256", secret);
      hmac.update(text);
      return hmac.digest("base64url");
    },
    async verify(text, signature, secret) {
      const expected = await this.sign(text, secret);
      const a = Buffer$1.from(expected);
      const b = Buffer$1.from(signature);
      return a.length === b.length && crypto.timingSafeEqual(a, b);
    }
  };
  encoding = {
    base64Encode: (text) => {
      return Buffer$1.from(text).toString("base64url");
    },
    base64Decode: (text) => {
      return Buffer$1.from(text, "base64url").toString("utf-8");
    }
  };
  getEnv(key) {
    return process.env[key];
  }
  async readFile(path2) {
    return readFile(path2, "utf-8");
  }
  async writeFile(path2, content) {
    await writeFile(path2, content, "utf-8");
  }
  async deleteFile(path2) {
    await rm(path2, { force: true });
  }
}
const SESSION_SCHEMA = {
  domain: "sessions",
  description: "A Jules session representing a task or conversation with the agent",
  fields: [
    {
      name: "id",
      type: "string",
      description: "Unique session identifier",
      filterable: true,
      selectable: true
    },
    {
      name: "name",
      type: "string",
      description: 'Full resource name (e.g., "sessions/314159...")',
      selectable: true
    },
    {
      name: "title",
      type: "string",
      description: "Human-readable session title",
      filterable: true,
      selectable: true
    },
    {
      name: "prompt",
      type: "string",
      description: "The initial instruction or task description",
      selectable: true
    },
    {
      name: "state",
      type: "SessionState",
      description: 'Current session state: "queued" | "planning" | "awaitingPlanApproval" | "awaitingUserFeedback" | "inProgress" | "paused" | "failed" | "completed"',
      filterable: true,
      selectable: true
    },
    {
      name: "createTime",
      type: "string",
      description: "RFC 3339 timestamp when session was created",
      filterable: true,
      selectable: true
    },
    {
      name: "updateTime",
      type: "string",
      description: "RFC 3339 timestamp when session was last updated",
      filterable: true,
      selectable: true
    },
    {
      name: "url",
      type: "string",
      description: "URL to view the session in the Jules web app",
      selectable: true
    },
    {
      name: "durationMs",
      type: "number",
      description: "Duration in milliseconds (updateTime - createTime)",
      computed: true,
      selectable: true
    },
    {
      name: "sourceContext",
      type: "SourceContext",
      description: "The source code context (repository, branch)",
      selectable: true,
      fields: [
        {
          name: "source",
          type: "string",
          description: 'Source name (e.g., "sources/github/owner/repo")',
          selectable: true
        },
        {
          name: "githubRepoContext",
          type: "object",
          description: "GitHub-specific context",
          optional: true,
          selectable: true,
          fields: [
            {
              name: "startingBranch",
              type: "string",
              description: "The branch the session started from",
              selectable: true
            }
          ]
        }
      ]
    },
    {
      name: "outputs",
      type: "SessionOutput[]",
      description: "Outputs generated by the session (e.g., pull requests)",
      selectable: true,
      fields: [
        {
          name: "type",
          type: '"pullRequest"',
          description: "Output type discriminator",
          selectable: true
        },
        {
          name: "pullRequest",
          type: "PullRequest",
          description: "Pull request details",
          selectable: true,
          fields: [
            {
              name: "url",
              type: "string",
              description: "PR URL",
              selectable: true
            },
            {
              name: "title",
              type: "string",
              description: "PR title",
              selectable: true
            },
            {
              name: "description",
              type: "string",
              description: "PR description",
              selectable: true
            }
          ]
        }
      ]
    },
    {
      name: "outcome",
      type: "object",
      description: "Final outcome if session is in terminal state",
      optional: true,
      selectable: true,
      fields: [
        {
          name: "status",
          type: '"SUCCESS" | "FAILURE"',
          description: "Outcome status",
          selectable: true
        },
        {
          name: "summary",
          type: "string",
          description: "Human-readable outcome summary",
          selectable: true
        }
      ]
    }
  ],
  examples: [
    {
      description: "Get all completed sessions",
      query: {
        from: "sessions",
        where: { state: "completed" },
        select: ["id", "title", "createTime"]
      }
    },
    {
      description: "Find sessions with PR output",
      query: {
        from: "sessions",
        where: { "outputs.type": "pullRequest" },
        select: ["id", "title", "outputs.pullRequest.url"]
      }
    },
    {
      description: "Get session with all activities",
      query: {
        from: "sessions",
        where: { id: "SESSION_ID" },
        include: { activities: true }
      }
    }
  ]
};
const ACTIVITY_SCHEMA = {
  domain: "activities",
  description: "An event or action within a session",
  fields: [
    {
      name: "id",
      type: "string",
      description: "Unique activity identifier",
      filterable: true,
      selectable: true
    },
    {
      name: "name",
      type: "string",
      description: "Full resource name",
      selectable: true
    },
    {
      name: "type",
      type: "ActivityType",
      description: 'Activity type: "agentMessaged" | "userMessaged" | "planGenerated" | "planApproved" | "progressUpdated" | "sessionCompleted" | "sessionFailed"',
      filterable: true,
      selectable: true
    },
    {
      name: "createTime",
      type: "string",
      description: "RFC 3339 timestamp when activity was created",
      filterable: true,
      selectable: true
    },
    {
      name: "originator",
      type: '"user" | "agent" | "system"',
      description: "Entity that originated this activity",
      filterable: true,
      selectable: true
    },
    {
      name: "sessionId",
      type: "string",
      description: "ID of the parent session",
      filterable: true,
      selectable: true
    },
    {
      name: "artifactCount",
      type: "number",
      description: "Number of artifacts in this activity",
      computed: true,
      selectable: true
    },
    {
      name: "summary",
      type: "string",
      description: "Human-readable summary of the activity",
      computed: true,
      selectable: true
    },
    // Type-specific fields
    {
      name: "message",
      type: "string",
      description: "Message content (for agentMessaged, userMessaged)",
      optional: true,
      selectable: true
    },
    {
      name: "plan",
      type: "Plan",
      description: "Generated plan (for planGenerated)",
      optional: true,
      selectable: true,
      fields: [
        {
          name: "id",
          type: "string",
          description: "Plan ID",
          selectable: true
        },
        {
          name: "createTime",
          type: "string",
          description: "Plan creation time",
          selectable: true
        },
        {
          name: "steps",
          type: "PlanStep[]",
          description: "Plan steps",
          selectable: true,
          fields: [
            {
              name: "id",
              type: "string",
              description: "Step ID",
              selectable: true
            },
            {
              name: "title",
              type: "string",
              description: "Step title",
              selectable: true
            },
            {
              name: "description",
              type: "string",
              description: "Step description",
              optional: true,
              selectable: true
            },
            {
              name: "index",
              type: "number",
              description: "Step index",
              selectable: true
            }
          ]
        }
      ]
    },
    {
      name: "planId",
      type: "string",
      description: "ID of approved plan (for planApproved)",
      optional: true,
      selectable: true
    },
    {
      name: "title",
      type: "string",
      description: "Progress title (for progressUpdated)",
      optional: true,
      selectable: true
    },
    {
      name: "description",
      type: "string",
      description: "Progress description (for progressUpdated)",
      optional: true,
      selectable: true
    },
    {
      name: "reason",
      type: "string",
      description: "Failure reason (for sessionFailed)",
      optional: true,
      selectable: true
    },
    {
      name: "artifacts",
      type: "Artifact[]",
      description: "Artifacts produced by this activity",
      selectable: true,
      fields: [
        {
          name: "type",
          type: '"changeSet" | "media" | "bashOutput"',
          description: "Artifact type discriminator",
          filterable: true,
          selectable: true
        },
        {
          name: "changeSet",
          type: "ChangeSet",
          description: "Code changes (for changeSet type)",
          optional: true,
          selectable: true,
          fields: [
            {
              name: "source",
              type: "string",
              description: "Source name",
              selectable: true
            },
            {
              name: "gitPatch",
              type: "GitPatch",
              description: "Git patch data",
              selectable: true,
              fields: [
                {
                  name: "unidiffPatch",
                  type: "string",
                  description: "Unidiff patch content",
                  selectable: true
                },
                {
                  name: "baseCommitId",
                  type: "string",
                  description: "Base commit SHA",
                  selectable: true
                },
                {
                  name: "suggestedCommitMessage",
                  type: "string",
                  description: "Suggested commit message",
                  selectable: true
                }
              ]
            }
          ]
        },
        {
          name: "command",
          type: "string",
          description: "Bash command executed (for bashOutput type)",
          optional: true,
          filterable: true,
          selectable: true
        },
        {
          name: "stdout",
          type: "string",
          description: "Standard output (for bashOutput type)",
          optional: true,
          selectable: true
        },
        {
          name: "stderr",
          type: "string",
          description: "Standard error (for bashOutput type)",
          optional: true,
          selectable: true
        },
        {
          name: "exitCode",
          type: "number | null",
          description: "Exit code (for bashOutput type)",
          optional: true,
          filterable: true,
          selectable: true
        },
        {
          name: "data",
          type: "string",
          description: "Base64 media data (for media type) - typically excluded",
          optional: true,
          selectable: true
        },
        {
          name: "format",
          type: "string",
          description: "Media format (for media type)",
          optional: true,
          selectable: true
        }
      ]
    }
  ],
  examples: [
    {
      description: "Get all plan generations from a session",
      query: {
        from: "activities",
        where: { sessionId: "SESSION_ID", type: "planGenerated" },
        select: ["id", "createTime", "plan.steps.title"]
      }
    },
    {
      description: "Find activities with bash commands that failed",
      query: {
        from: "activities",
        where: {
          "artifacts.type": "bashOutput",
          "artifacts.exitCode": { neq: 0 }
        },
        select: [
          "id",
          "artifacts.command",
          "artifacts.exitCode",
          "artifacts.stderr"
        ]
      }
    },
    {
      description: "Get agent messages with their artifacts",
      query: {
        from: "activities",
        where: { type: "agentMessaged" },
        select: ["id", "message", "artifactCount", "artifacts.type"]
      }
    },
    {
      description: "Find activities with code changes",
      query: {
        from: "activities",
        where: { "artifacts.type": "changeSet" },
        select: [
          "id",
          "createTime",
          "artifacts.changeSet.gitPatch.suggestedCommitMessage"
        ]
      }
    }
  ]
};
const FILTER_OP_SCHEMA = {
  description: "Filter operators for where clause",
  operators: [
    {
      name: "eq",
      description: "Equals (also supports direct value)",
      example: '{ id: "abc" } or { id: { eq: "abc" } }'
    },
    {
      name: "neq",
      description: "Not equals",
      example: '{ state: { neq: "failed" } }'
    },
    {
      name: "contains",
      description: "Case-insensitive substring match",
      example: '{ title: { contains: "bug" } }'
    },
    {
      name: "gt",
      description: "Greater than",
      example: '{ createTime: { gt: "2024-01-01" } }'
    },
    {
      name: "lt",
      description: "Less than",
      example: '{ createTime: { lt: "2024-12-31" } }'
    },
    {
      name: "gte",
      description: "Greater than or equal",
      example: "{ artifactCount: { gte: 1 } }"
    },
    {
      name: "lte",
      description: "Less than or equal",
      example: "{ artifactCount: { lte: 10 } }"
    },
    {
      name: "in",
      description: "Value in array",
      example: '{ state: { in: ["completed", "failed"] } }'
    },
    {
      name: "exists",
      description: "Field existence check",
      example: '{ "outputs.pullRequest": { exists: true } }'
    }
  ],
  dotNotation: {
    description: "Use dot notation for nested field paths. When filtering arrays, uses existential matching (ANY element matches).",
    examples: [
      '"artifacts.type": "bashOutput"',
      '"artifacts.exitCode": { neq: 0 }',
      '"plan.steps.title": { contains: "test" }',
      '"outputs.pullRequest.url": { exists: true }'
    ]
  }
};
const PROJECTION_SCHEMA = {
  description: "Select clause projection syntax",
  syntax: [
    {
      name: "Field selection",
      description: "Include specific fields",
      example: '["id", "title", "state"]'
    },
    {
      name: "Dot notation",
      description: "Select nested fields",
      example: '["id", "artifacts.type", "artifacts.command"]'
    },
    {
      name: "Wildcard",
      description: "Include all fields",
      example: '["*"]'
    },
    {
      name: "Exclusion",
      description: "Exclude specific fields (use with wildcard)",
      example: '["*", "-artifacts.data"]'
    }
  ],
  defaults: {
    sessions: ["id", "state", "title", "createTime"],
    activities: [
      "id",
      "type",
      "createTime",
      "originator",
      "artifactCount",
      "summary"
    ]
  }
};
function getSchema(domain) {
  return domain === "sessions" ? SESSION_SCHEMA : ACTIVITY_SCHEMA;
}
function getAllSchemas() {
  return {
    sessions: SESSION_SCHEMA,
    activities: ACTIVITY_SCHEMA,
    filterOps: FILTER_OP_SCHEMA,
    projection: PROJECTION_SCHEMA
  };
}
function generateTypeDefinition(domain) {
  const schema = getSchema(domain);
  const lines = [];
  lines.push(`/**`);
  lines.push(` * ${schema.description}`);
  lines.push(` */`);
  const typeName = domain === "sessions" ? "SessionResource" : "Activity";
  lines.push(`interface ${typeName} {`);
  for (const field of schema.fields) {
    const optional = field.optional ? "?" : "";
    const computed = field.computed ? " // computed" : "";
    lines.push(`  /** ${field.description} */`);
    lines.push(`  ${field.name}${optional}: ${field.type};${computed}`);
  }
  lines.push(`}`);
  return lines.join("\n");
}
function generateMarkdownDocs() {
  const lines = [];
  lines.push("# Jules Query Language (JQL) Schema Reference\n");
  lines.push("## Sessions Domain\n");
  lines.push(SESSION_SCHEMA.description + "\n");
  lines.push("### Fields\n");
  lines.push("| Field | Type | Description | Filterable |");
  lines.push("|-------|------|-------------|------------|");
  for (const field of SESSION_SCHEMA.fields) {
    const filterable = field.filterable ? "Yes" : "No";
    const computed = field.computed ? " (computed)" : "";
    lines.push(
      `| ${field.name} | ${field.type}${computed} | ${field.description} | ${filterable} |`
    );
  }
  lines.push("\n## Activities Domain\n");
  lines.push(ACTIVITY_SCHEMA.description + "\n");
  lines.push("### Fields\n");
  lines.push("| Field | Type | Description | Filterable |");
  lines.push("|-------|------|-------------|------------|");
  for (const field of ACTIVITY_SCHEMA.fields) {
    const filterable = field.filterable ? "Yes" : "No";
    const computed = field.computed ? " (computed)" : "";
    lines.push(
      `| ${field.name} | ${field.type}${computed} | ${field.description} | ${filterable} |`
    );
  }
  lines.push("\n## Filter Operators\n");
  for (const op of FILTER_OP_SCHEMA.operators) {
    lines.push(`- **${op.name}**: ${op.description}`);
    lines.push(`  - Example: \`${op.example}\``);
  }
  lines.push("\n### Dot Notation\n");
  lines.push(FILTER_OP_SCHEMA.dotNotation.description + "\n");
  lines.push("Examples:");
  for (const ex of FILTER_OP_SCHEMA.dotNotation.examples) {
    lines.push(`- \`${ex}\``);
  }
  lines.push("\n## Projection (Select)\n");
  for (const syntax of PROJECTION_SCHEMA.syntax) {
    lines.push(`- **${syntax.name}**: ${syntax.description}`);
    lines.push(`  - Example: \`${syntax.example}\``);
  }
  lines.push("\n## Query Examples\n");
  lines.push("### Sessions");
  for (const ex of SESSION_SCHEMA.examples) {
    lines.push(`
**${ex.description}**`);
    lines.push("```json");
    lines.push(JSON.stringify(ex.query, null, 2));
    lines.push("```");
  }
  lines.push("\n### Activities");
  for (const ex of ACTIVITY_SCHEMA.examples) {
    lines.push(`
**${ex.description}**`);
    lines.push("```json");
    lines.push(JSON.stringify(ex.query, null, 2));
    lines.push("```");
  }
  return lines.join("\n");
}
const VALID_OPERATORS = new Set(
  FILTER_OP_SCHEMA.operators.map((op) => op.name)
);
const VALID_DOMAINS = /* @__PURE__ */ new Set(["sessions", "activities"]);
const VALID_ORDERS = /* @__PURE__ */ new Set(["asc", "desc"]);
function resolveFieldPath(path2, domain) {
  const schema = domain === "sessions" ? SESSION_SCHEMA : ACTIVITY_SCHEMA;
  const parts = path2.split(".");
  let currentFields = schema.fields;
  let currentField = null;
  for (const part of parts) {
    const found = currentFields.find((f) => f.name === part);
    if (!found) {
      return { field: null, exists: false, computedField: false };
    }
    currentField = found;
    currentFields = found.fields || [];
  }
  return {
    field: currentField,
    exists: true,
    computedField: currentField?.computed || false
  };
}
function getValidFieldPaths(domain, prefix = "") {
  const schema = domain === "sessions" ? SESSION_SCHEMA : ACTIVITY_SCHEMA;
  const paths = [];
  function collectPaths(fields, currentPrefix) {
    for (const field of fields) {
      const fullPath = currentPrefix ? `${currentPrefix}.${field.name}` : field.name;
      paths.push(fullPath);
      if (field.fields) {
        collectPaths(field.fields, fullPath);
      }
    }
  }
  collectPaths(schema.fields, prefix);
  return paths;
}
function validateStructure(query, errors) {
  if (typeof query !== "object" || query === null || Array.isArray(query)) {
    errors.push({
      code: "INVALID_STRUCTURE",
      path: "",
      message: "Query must be a non-null object",
      suggestion: 'Provide a query object like { from: "sessions" }'
    });
    return false;
  }
  return true;
}
function validateDomain(query, errors) {
  if (!("from" in query)) {
    errors.push({
      code: "MISSING_REQUIRED_FIELD",
      path: "from",
      message: "Missing required field: from",
      suggestion: 'Add from: "sessions" or from: "activities"'
    });
    return null;
  }
  const from = query.from;
  if (typeof from !== "string" || !VALID_DOMAINS.has(from)) {
    errors.push({
      code: "INVALID_DOMAIN",
      path: "from",
      message: `Invalid domain: "${from}"`,
      suggestion: 'Valid domains are: "sessions", "activities"'
    });
    return null;
  }
  return from;
}
function validateSelect(select2, domain, errors, warnings) {
  if (select2 === void 0) return;
  if (!Array.isArray(select2)) {
    errors.push({
      code: "INVALID_STRUCTURE",
      path: "select",
      message: "select must be an array of strings",
      suggestion: 'Use select: ["id", "title"] or omit for default projection'
    });
    return;
  }
  for (let i = 0; i < select2.length; i++) {
    const expr = select2[i];
    if (typeof expr !== "string") {
      errors.push({
        code: "INVALID_SELECT_EXPRESSION",
        path: `select[${i}]`,
        message: `Select expression must be a string, got ${typeof expr}`
      });
      continue;
    }
    if (expr === "*") continue;
    const isExclusion = expr.startsWith("-");
    const fieldPath = isExclusion ? expr.slice(1) : expr;
    const resolution = resolveFieldPath(fieldPath, domain);
    if (!resolution.exists) {
      const validPaths = getValidFieldPaths(domain);
      const similar = validPaths.filter(
        (p) => p.includes(fieldPath.split(".")[0]) || fieldPath.split(".")[0].includes(p.split(".")[0])
      );
      warnings.push({
        code: "UNKNOWN_FIELD",
        path: `select[${i}]`,
        message: `Unknown field path: "${fieldPath}"`
      });
      if (similar.length > 0) {
        warnings[warnings.length - 1].message += `. Did you mean: ${similar.slice(0, 3).join(", ")}?`;
      }
    }
  }
}
function validateWhere(where, domain, errors, warnings) {
  if (where === void 0) return;
  if (typeof where !== "object" || where === null || Array.isArray(where)) {
    errors.push({
      code: "INVALID_STRUCTURE",
      path: "where",
      message: "where must be an object",
      suggestion: 'Use where: { field: "value" } or where: { field: { op: value } }'
    });
    return;
  }
  const whereObj = where;
  for (const [key, value] of Object.entries(whereObj)) {
    const fieldPath = `where.${key}`;
    if (key === "search") {
      if (typeof value !== "string") {
        errors.push({
          code: "INVALID_OPERATOR_VALUE",
          path: fieldPath,
          message: "search must be a string"
        });
      }
      continue;
    }
    const resolution = resolveFieldPath(key, domain);
    if (!resolution.exists) {
      getValidFieldPaths(domain).filter(
        (p) => SESSION_SCHEMA.fields.find((f) => f.name === p)?.filterable || ACTIVITY_SCHEMA.fields.find((f) => f.name === p)?.filterable
      );
      warnings.push({
        code: "UNKNOWN_FIELD",
        path: fieldPath,
        message: `Unknown field: "${key}"`
      });
    } else if (resolution.computedField) {
      errors.push({
        code: "COMPUTED_FIELD_FILTER",
        path: fieldPath,
        message: `Cannot filter on computed field: "${key}"`,
        suggestion: "Computed fields (artifactCount, summary, durationMs) can only be selected, not filtered"
      });
    } else if (resolution.field && !resolution.field.filterable) {
      warnings.push({
        code: "NON_FILTERABLE_FIELD",
        path: fieldPath,
        message: `Field "${key}" may not be efficiently filterable`
      });
    }
    validateFilterValue(value, fieldPath, errors);
  }
}
function validateFilterValue(value, path2, errors) {
  if (value === null || typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
    return;
  }
  if (typeof value === "object" && !Array.isArray(value)) {
    const opObj = value;
    const keys = Object.keys(opObj);
    if (keys.length === 0) {
      errors.push({
        code: "INVALID_OPERATOR",
        path: path2,
        message: "Empty filter object",
        suggestion: "Use { eq: value } or a direct value"
      });
      return;
    }
    for (const op of keys) {
      if (!VALID_OPERATORS.has(op)) {
        errors.push({
          code: "INVALID_OPERATOR",
          path: `${path2}.${op}`,
          message: `Invalid operator: "${op}"`,
          suggestion: `Valid operators: ${Array.from(VALID_OPERATORS).join(", ")}`
        });
        continue;
      }
      const opValue = opObj[op];
      switch (op) {
        case "eq":
        case "neq":
        case "gt":
        case "lt":
        case "gte":
        case "lte":
          if (opValue !== null && typeof opValue !== "string" && typeof opValue !== "number" && typeof opValue !== "boolean") {
            errors.push({
              code: "INVALID_OPERATOR_VALUE",
              path: `${path2}.${op}`,
              message: `${op} operator requires a primitive value (string, number, boolean, or null)`
            });
          }
          break;
        case "contains":
          if (typeof opValue !== "string") {
            errors.push({
              code: "INVALID_OPERATOR_VALUE",
              path: `${path2}.${op}`,
              message: "contains operator requires a string value"
            });
          }
          break;
        case "in":
          if (!Array.isArray(opValue)) {
            errors.push({
              code: "INVALID_OPERATOR_VALUE",
              path: `${path2}.${op}`,
              message: "in operator requires an array of values",
              suggestion: 'Use in: ["value1", "value2"]'
            });
          }
          break;
        case "exists":
          if (typeof opValue !== "boolean") {
            errors.push({
              code: "INVALID_OPERATOR_VALUE",
              path: `${path2}.${op}`,
              message: "exists operator requires a boolean value",
              suggestion: "Use exists: true or exists: false"
            });
          }
          break;
      }
    }
  } else {
    errors.push({
      code: "INVALID_OPERATOR_VALUE",
      path: path2,
      message: `Invalid filter value type: ${typeof value}`,
      suggestion: 'Use a primitive value for equality or an operator object like { contains: "text" }'
    });
  }
}
function validateOrder(order, errors) {
  if (order === void 0) return;
  if (typeof order !== "string" || !VALID_ORDERS.has(order)) {
    errors.push({
      code: "INVALID_ORDER",
      path: "order",
      message: `Invalid order: "${order}"`,
      suggestion: 'Valid values are: "asc", "desc"'
    });
  }
}
function validateLimit(limit, errors, warnings) {
  if (limit === void 0) return;
  if (typeof limit !== "number" || !Number.isInteger(limit)) {
    errors.push({
      code: "INVALID_LIMIT",
      path: "limit",
      message: "limit must be an integer"
    });
    return;
  }
  if (limit < 0) {
    errors.push({
      code: "INVALID_LIMIT",
      path: "limit",
      message: "limit cannot be negative"
    });
  }
  if (limit > 1e3) {
    warnings.push({
      code: "LIMIT_TOO_HIGH",
      path: "limit",
      message: `limit of ${limit} exceeds maximum of 1000, will be capped`
    });
  }
}
function validateCursor(cursor, field, errors) {
  if (cursor === void 0) return;
  if (typeof cursor !== "string") {
    errors.push({
      code: "INVALID_STRUCTURE",
      path: field,
      message: `${field} must be a string (ID)`
    });
  }
}
function validateQuery(query) {
  const errors = [];
  const warnings = [];
  if (!validateStructure(query, errors)) {
    return { valid: false, errors, warnings };
  }
  const queryObj = query;
  const domain = validateDomain(queryObj, errors);
  if (!domain) {
    return { valid: false, errors, warnings };
  }
  validateSelect(queryObj.select, domain, errors, warnings);
  validateWhere(queryObj.where, domain, errors, warnings);
  validateOrder(queryObj.order, errors);
  validateLimit(queryObj.limit, errors, warnings);
  validateCursor(queryObj.startAfter, "startAfter", errors);
  validateCursor(queryObj.startAt, "startAt", errors);
  const validTopLevelFields = /* @__PURE__ */ new Set([
    "from",
    "select",
    "where",
    "order",
    "limit",
    "startAfter",
    "startAt",
    "include",
    "tokenBudget",
    "offset"
  ]);
  for (const key of Object.keys(queryObj)) {
    if (!validTopLevelFields.has(key)) {
      warnings.push({
        code: "UNKNOWN_QUERY_FIELD",
        path: key,
        message: `Unknown query field: "${key}"`
      });
    }
  }
  return {
    valid: errors.length === 0,
    errors,
    warnings
  };
}
function formatValidationResult(result) {
  const lines = [];
  if (result.valid) {
    lines.push("Query is valid.");
  } else {
    lines.push("Query validation failed:");
  }
  if (result.errors.length > 0) {
    lines.push("\nErrors:");
    for (const error of result.errors) {
      lines.push(`  - [${error.code}] ${error.path}: ${error.message}`);
      if (error.suggestion) {
        lines.push(`    Suggestion: ${error.suggestion}`);
      }
    }
  }
  if (result.warnings.length > 0) {
    lines.push("\nWarnings:");
    for (const warning of result.warnings) {
      lines.push(`  - [${warning.code}] ${warning.path}: ${warning.message}`);
    }
  }
  return lines.join("\n");
}
class MemoryStorage {
  activities = [];
  indices = /* @__PURE__ */ new Map();
  /**
   * Initializes the storage. No-op for memory storage.
   */
  async init() {
  }
  /**
   * Closes the storage and clears memory.
   */
  async close() {
    this.activities = [];
    this.indices.clear();
  }
  /**
   * Appends an activity to the in-memory list.
   *
   * **Guarantee:**
   * - Idempotent: If an activity with the same ID exists, it updates it in place.
   * - Append-only: New activities are always added to the end.
   *
   * **Side Effects:**
   * - Modifies the internal `activities` array.
   */
  async append(activity) {
    if (this.indices.has(activity.id)) {
      const index = this.indices.get(activity.id);
      this.activities[index] = activity;
    } else {
      const index = this.activities.push(activity) - 1;
      this.indices.set(activity.id, index);
    }
  }
  /**
   * Retrieves an activity by ID.
   */
  async get(activityId) {
    const index = this.indices.get(activityId);
    if (index !== void 0) {
      return this.activities[index];
    }
    return void 0;
  }
  /**
   * Retrieves the latest activity.
   */
  async latest() {
    if (this.activities.length === 0) return void 0;
    return this.activities[this.activities.length - 1];
  }
  /**
   * Yields all activities in chronological order.
   */
  async *scan() {
    for (const activity of this.activities) {
      yield activity;
    }
  }
}
class MemorySessionStorage {
  sessions = /* @__PURE__ */ new Map();
  index = [];
  async init() {
  }
  async upsert(session) {
    this.sessions.set(session.id, {
      resource: session,
      _lastSyncedAt: Date.now()
    });
    this.index.push({
      id: session.id,
      title: session.title,
      state: session.state,
      createTime: session.createTime,
      source: session.sourceContext?.source || "unknown",
      _updatedAt: Date.now()
    });
  }
  async upsertMany(sessions) {
    for (const s of sessions) {
      await this.upsert(s);
    }
  }
  async get(sessionId) {
    return this.sessions.get(sessionId);
  }
  async delete(sessionId) {
    this.sessions.delete(sessionId);
  }
  async *scanIndex() {
    for (const entry of this.index) {
      yield entry;
    }
  }
}
const defaultPlatform = new NodePlatform();
const defaultStorageFactory = {
  activity: (sessionId) => new NodeFileStorage(sessionId, getRootDir()),
  session: () => new NodeSessionStorage(getRootDir())
};
function connect(options = {}) {
  return new JulesClientImpl(options, defaultStorageFactory, defaultPlatform);
}
const jules = connect();
export {
  ACTIVITY_SCHEMA,
  AutomatedSessionFailedError,
  BashArtifact,
  ChangeSetArtifact,
  FILTER_OP_SCHEMA,
  InvalidStateError,
  JulesApiError,
  JulesAuthenticationError,
  JulesClientImpl,
  JulesError,
  JulesNetworkError,
  JulesRateLimitError,
  MemorySessionStorage,
  MemoryStorage,
  MissingApiKeyError,
  NodePlatform,
  PROJECTION_SCHEMA,
  SESSION_SCHEMA,
  SessionCursor,
  SourceNotFoundError,
  SyncInProgressError,
  TimeoutError,
  connect,
  formatValidationResult,
  generateMarkdownDocs,
  generateTypeDefinition,
  getAllSchemas,
  getSchema,
  jules,
  parseUnidiff,
  toSummary,
  validateQuery
};
